commit 5afe9215ea4b1a604f460c64395ff10dcfca8842
Author: Mingsheng Xu <mxu61@ucsc.edu>
Date:   Wed Jul 10 16:20:21 2024 -0700

    Implementing uop cache

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 0e9d428..693a988 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -3,7 +3,7 @@ cmake_minimum_required (VERSION 3.5)
 project(scarab C CXX ASM)
 
 set(CMAKE_C_STANDARD 99)
-set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_C_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
diff --git a/src/bp/bp.c b/src/bp/bp.c
index 38a24e6..c7c6bd9 100644
--- a/src/bp/bp.c
+++ b/src/bp/bp.c
@@ -148,14 +148,13 @@ void bp_sched_recovery(Bp_Recovery_Info* bp_recovery_info, Op* op,
      op->op_num <= bp_recovery_info->recovery_op_num) {
     const Addr next_fetch_addr = op->oracle_info.npc;
     ASSERT(0, op->oracle_info.npc);
-    Flag uc_hit = in_uop_cache(next_fetch_addr, FALSE, op->off_path);
     const uns latency = late_bp_recovery ? LATE_BP_LATENCY : 1;
     DEBUG(
       bp_recovery_info->proc_id,
       "Recovery signaled for op_num:%s @ 0x%s  next_fetch:0x%s offpath:%d\n",
       unsstr64(op->op_num), hexstr64s(op->inst_info->addr),
       hexstr64s(next_fetch_addr), op->off_path);
-    inc_bstat_miss(op, uc_hit);
+    inc_bstat_miss(op);
     ASSERT(op->proc_id, !op->oracle_info.recovery_sch);
     op->oracle_info.recovery_sch          = TRUE;
     bp_recovery_info->recovery_cycle      = cycle + latency;
@@ -198,28 +197,12 @@ void inc_bstat_fetched(Op* op) {
     bstat->cf_type = op->table_info->cf_type;
   }
 
-  if (!op->off_path) {
-    if (in_uop_cache(op->oracle_info.npc, FALSE, op->off_path)) {
-      bstat->bpu_hit_uc_hit_on_path += 1;
-    }else {
-      bstat->bpu_hit_uc_miss_on_path += 1;
-      if (!in_icache(op->oracle_info.npc)) bstat->bpu_hit_uc_ic_miss_on_path += 1;
-    }
-  }
-  else {
-    if (in_uop_cache(op->oracle_info.npc, FALSE, op->off_path)) {
-      bstat->bpu_hit_uc_hit_off_path += 1;
-    }else {
-      bstat->bpu_hit_uc_miss_off_path += 1;
-      if (!in_icache(op->oracle_info.npc)) bstat->bpu_hit_uc_ic_miss_off_path += 1;
-    }
-  }
   // target if taken
   if (op->oracle_info.pred == TAKEN && !(op->oracle_info.recover_at_exec || op->oracle_info.recover_at_decode))
     bstat->target = op->oracle_info.npc;
 }
 
-void inc_bstat_miss(Op* op, Flag uc_hit) {
+void inc_bstat_miss(Op* op) {
   int64 key = convert_to_cmp_addr(op->table_info->cf_type, op->inst_info->addr);
   Per_Branch_Stat* bstat = (Per_Branch_Stat*) hash_table_access(&per_branch_stat, key);
   ASSERT(bp_recovery_info->proc_id, bstat);
@@ -239,34 +222,8 @@ void inc_bstat_miss(Op* op, Flag uc_hit) {
     return;
   }
 
-  const Flag in_ic = in_icache(op->oracle_info.npc);
-
   if (op->fetched_from_uop_cache && op->oracle_info.recover_at_decode)
     STAT_EVENT(bp_recovery_info->proc_id, RECOVER_AT_DECODE_BR_FROM_UOC);
-
-  if (uc_hit) {
-    if (mispred)        bstat->mispred_uc_hit += 1;
-    else if (misfetch)  bstat->misfetch_uc_hit += 1;
-    else if (btb_miss)  bstat->btb_miss_uc_hit += 1;
-    else                bstat->other_recovery_uc_hit += 1;
-    STAT_EVENT(bp_recovery_info->proc_id, RECOVERED_FROM_UOC);
-  } else {
-    if (mispred) {
-      bstat->mispred_uc_miss += 1;
-      if (!in_ic) bstat->mispred_uc_ic_miss += 1;
-    } else if (misfetch) {
-      bstat->misfetch_uc_miss += 1;
-      if (!in_ic) bstat->misfetch_uc_ic_miss += 1;
-    } else if (btb_miss){
-      bstat->btb_miss_uc_miss += 1;
-      if (!in_ic) bstat->btb_miss_uc_ic_miss += 1;
-    }
-    else {
-      bstat->other_recovery_uc_miss += 1;
-      if (!in_ic) bstat->other_recovery_uc_ic_miss += 1;
-    }
-    bstat->recover_redirect_extra_fetch_latency += ICACHE_LATENCY - UOP_CACHE_LATENCY;
-  }
 }
 
 /******************************************************************************/
@@ -280,8 +237,7 @@ void bp_sched_redirect(Bp_Recovery_Info* bp_recovery_info, Op* op,
     DEBUG(bp_recovery_info->proc_id, "Redirect signaled for op_num:%s @ 0x%s\n",
           unsstr64(op->op_num), hexstr64s(op->inst_info->addr));
 
-    Flag uc_hit = in_uop_cache(op->oracle_info.npc, FALSE, op->off_path);
-    inc_bstat_miss(op, uc_hit); // shouldn't double count if both btb miss and predictor wrong.
+    inc_bstat_miss(op); // shouldn't double count if both btb miss and predictor wrong.
     bp_recovery_info->redirect_cycle = cycle + 1 +
                                        (op->table_info->cf_type == CF_SYS ?
                                           EXTRA_CALLSYS_CYCLES :
diff --git a/src/bp/bp.h b/src/bp/bp.h
index d7ca7de..2fc7592 100644
--- a/src/bp/bp.h
+++ b/src/bp/bp.h
@@ -287,7 +287,7 @@ void bp_retire_op(Bp_Data*, Op*);
 void bp_recover_op(Bp_Data*, Cf_Type, Recovery_Info*);
 
 void inc_bstat_fetched(Op* op);
-void inc_bstat_miss(Op* op, Flag uc_hit);
+void inc_bstat_miss(Op* op);
 
 /**************************************************************************************/
 
diff --git a/src/cmp_model_support.c b/src/cmp_model_support.c
index 3a7e661..3a9ae92 100644
--- a/src/cmp_model_support.c
+++ b/src/cmp_model_support.c
@@ -65,6 +65,7 @@ void cmp_init_cmp_model() {
   alloc_mem_eip(NUM_CORES);
   alloc_mem_djolt(NUM_CORES);
   alloc_mem_fnlmma(NUM_CORES);
+  alloc_mem_uop_cache(NUM_CORES);
 }
 
 
@@ -86,6 +87,7 @@ void cmp_set_all_stages(uns8 proc_id) {
   set_fnlmma(proc_id);
   set_fdip(proc_id, &cmp_model.icache_stage[proc_id]);
   set_decoupled_fe(proc_id);
+  set_uop_cache(proc_id);
   set_icache_stage(&cmp_model.icache_stage[proc_id]);
   set_decode_stage(&cmp_model.decode_stage[proc_id]);
   set_map_stage(&cmp_model.map_stage[proc_id]);
diff --git a/src/core.stat.def b/src/core.stat.def
index fac22c2..e0dc6d0 100644
--- a/src/core.stat.def
+++ b/src/core.stat.def
@@ -466,9 +466,6 @@ DEF_STAT(  WRONG_IO_SCHED,     COUNT,  NO_RATIO    )
 
 DEF_STAT(  DVFS_CONFIG_SWITCH, COUNT,  NO_RATIO    )
 
-DEF_STAT(  UOP_CACHE_ICACHE_SWITCH_ONPATH,   DIST, NO_RATIO)
-DEF_STAT(  UOP_CACHE_ICACHE_SWITCH_OFFPATH,   DIST, NO_RATIO)
-
 DEF_STAT(  MAP_STAGE_RECEIVED_OPS_0,				DIST,  NO_RATIO)
 DEF_STAT(  MAP_STAGE_RECEIVED_OPS_1,				COUNT, NO_RATIO)
 DEF_STAT(  MAP_STAGE_RECEIVED_OPS_2,				COUNT, NO_RATIO)
diff --git a/src/debug/debug.param.def b/src/debug/debug.param.def
index 0eda28f..d2cad8a 100644
--- a/src/debug/debug.param.def
+++ b/src/debug/debug.param.def
@@ -84,6 +84,7 @@ DEF_PARAM(  debug_thread,          DEBUG_THREAD,          Flag,  Flag,  FALSE,
 DEF_PARAM(  debug_decoupled_fe,    DEBUG_DECOUPLED_FE,    Flag,  Flag,  FALSE,  )
 DEF_PARAM(  debug_icache_stage,    DEBUG_ICACHE_STAGE,    Flag,  Flag,  FALSE,  )
 DEF_PARAM(  debug_decode_stage,    DEBUG_DECODE_STAGE,    Flag,  Flag,  FALSE,  )
+DEF_PARAM(  debug_cpp_cache,       DEBUG_CPP_CACHE,       Flag,  Flag,  FALSE,  )
 DEF_PARAM(  debug_uop_cache,       DEBUG_UOP_CACHE,       Flag,  Flag,  FALSE,  )
 DEF_PARAM(  debug_uop_queue_stage, DEBUG_UOP_QUEUE_STAGE, Flag,  Flag,  FALSE,  )
 DEF_PARAM(  debug_map_stage,       DEBUG_MAP_STAGE,       Flag,  Flag,  FALSE,  )
diff --git a/src/decoupled_frontend.cc b/src/decoupled_frontend.cc
index 6491177..08d5757 100644
--- a/src/decoupled_frontend.cc
+++ b/src/decoupled_frontend.cc
@@ -27,6 +27,7 @@ typedef struct FT_struct {
   void ft_free_ops_and_clear();
   bool ft_can_fetch_op();
   Op* ft_fetch_op();
+  void ft_set_per_op_ft_info();
 } FT;
 
 // Per core fetch target queue:
@@ -301,17 +302,19 @@ void update_decoupled_fe() {
     // ft_ended_by != FT_NOT_ENDED indicates the end of the current fetch target
     // it is now ready to be pushed to the queue
     if (ft_ended_by != FT_NOT_ENDED) {
-      ASSERT(set_proc_id, per_core_current_ft_to_push[set_proc_id].ft_info.start && per_core_current_ft_to_push[set_proc_id].ft_info.length && per_core_current_ft_to_push[set_proc_id].ops.size());
+      ASSERT(set_proc_id, per_core_current_ft_to_push[set_proc_id].ft_info.static_info.start && per_core_current_ft_to_push[set_proc_id].ft_info.static_info.length && per_core_current_ft_to_push[set_proc_id].ops.size());
+      ASSERT(set_proc_id, per_core_current_ft_to_push[set_proc_id].ops.front()->bom && per_core_current_ft_to_push[set_proc_id].ops.back()->eom);
+      per_core_current_ft_to_push[set_proc_id].ft_set_per_op_ft_info();
       if (!df_ftq->empty()) {
         // sanity check of consecutivity
         Op* last_op = df_ftq->back().ops.back();
-        if (df_ftq->back().ft_info.ended_by == FT_TAKEN_BRANCH) {
-          ASSERT(set_proc_id, last_op->oracle_info.pred_npc == per_core_current_ft_to_push[set_proc_id].ft_info.start);
-        } else if (df_ftq->back().ft_info.ended_by == FT_BAR_FETCH) {
-          ASSERT(set_proc_id, last_op->oracle_info.pred_npc == per_core_current_ft_to_push[set_proc_id].ft_info.start ||
-                              last_op->inst_info->addr + last_op->inst_info->trace_info.inst_size == per_core_current_ft_to_push[set_proc_id].ft_info.start);
+        if (df_ftq->back().ft_info.dynamic_info.ended_by == FT_TAKEN_BRANCH) {
+          ASSERT(set_proc_id, last_op->oracle_info.pred_npc == per_core_current_ft_to_push[set_proc_id].ft_info.static_info.start);
+        } else if (df_ftq->back().ft_info.dynamic_info.ended_by == FT_BAR_FETCH) {
+          ASSERT(set_proc_id, last_op->oracle_info.pred_npc == per_core_current_ft_to_push[set_proc_id].ft_info.static_info.start ||
+                              last_op->inst_info->addr + last_op->inst_info->trace_info.inst_size == per_core_current_ft_to_push[set_proc_id].ft_info.static_info.start);
         } else {
-          ASSERT(set_proc_id, last_op->inst_info->addr + last_op->inst_info->trace_info.inst_size == per_core_current_ft_to_push[set_proc_id].ft_info.start);
+          ASSERT(set_proc_id, last_op->inst_info->addr + last_op->inst_info->trace_info.inst_size == per_core_current_ft_to_push[set_proc_id].ft_info.static_info.start);
         }
       }
       df_ftq->emplace_back(per_core_current_ft_to_push[set_proc_id]);
@@ -391,7 +394,7 @@ FT_Info decoupled_fe_peek_ft(int proc_id) {
     return FT_Info();
   }
 }
-  
+
 decoupled_fe_iter* decoupled_fe_new_ftq_iter() {
   per_core_ftq_iterators[set_proc_id].push_back(decoupled_fe_iter());
   return &per_core_ftq_iterators[set_proc_id].back();
@@ -498,8 +501,9 @@ uint64_t decoupled_fe_get_ftq_num(int proc_id) {
 
 void FT::ft_add_op(Op *op, FT_Ended_By ft_ended_by) {
   if (ops.empty()) {
-    ASSERT(set_proc_id, op->bom && !ft_info.start);
-    ft_info.start = op->inst_info->addr;
+    ASSERT(set_proc_id, op->bom && !ft_info.static_info.start);
+    ft_info.static_info.start = op->inst_info->addr;
+    ft_info.dynamic_info.first_op_off_path = op->off_path;
   } else {
     if (op->bom) {
       // assert consecutivity
@@ -512,11 +516,12 @@ void FT::ft_add_op(Op *op, FT_Ended_By ft_ended_by) {
   }
   ops.emplace_back(op);
   if (ft_ended_by != FT_NOT_ENDED) {
-    ASSERT(set_proc_id, op->eom && !ft_info.length);
-    ASSERT(set_proc_id, ft_info.start);
-    ft_info.length = op->inst_info->addr + op->inst_info->trace_info.inst_size - ft_info.start;
-    ASSERT(set_proc_id, ft_info.ended_by == FT_NOT_ENDED);
-    ft_info.ended_by = ft_ended_by;
+    ASSERT(set_proc_id, op->eom && !ft_info.static_info.length);
+    ASSERT(set_proc_id, ft_info.static_info.start);
+    ft_info.static_info.n_uops = ops.size();
+    ft_info.static_info.length = op->inst_info->addr + op->inst_info->trace_info.inst_size - ft_info.static_info.start;
+    ASSERT(set_proc_id, ft_info.dynamic_info.ended_by == FT_NOT_ENDED);
+    ft_info.dynamic_info.ended_by = ft_ended_by;
   }
 }
 
@@ -528,9 +533,11 @@ void FT::ft_free_ops_and_clear() {
 
   ops.clear();
   op_pos = 0;
-  ft_info.start = 0;
-  ft_info.length = 0;
-  ft_info.ended_by = FT_NOT_ENDED;
+  ft_info.static_info.start = 0;
+  ft_info.static_info.length = 0;
+  ft_info.static_info.n_uops = 0;
+  ft_info.dynamic_info.ended_by = FT_NOT_ENDED;
+  ft_info.dynamic_info.first_op_off_path = FALSE;
 }
 
 bool FT::ft_can_fetch_op() {
@@ -548,3 +555,9 @@ Op* FT::ft_fetch_op() {
 
   return op;
 }
+
+void FT::ft_set_per_op_ft_info() {
+  for (auto op : ops) {
+    op->ft_info = ft_info;
+  }
+}
\ No newline at end of file
diff --git a/src/decoupled_frontend.h b/src/decoupled_frontend.h
index 67a3dce..f1cc39c 100644
--- a/src/decoupled_frontend.h
+++ b/src/decoupled_frontend.h
@@ -64,23 +64,6 @@ extern "C" {
     uint64_t flattened_op_pos;
   };
 
-  typedef enum FT_Ended_By_enum {
-   FT_NOT_ENDED,
-   FT_ICACHE_LINE_BOUNDARY,
-   FT_TAKEN_BRANCH,
-   FT_BAR_FETCH,
-   FT_APP_EXIT
-  } FT_Ended_By;
-
-  typedef struct FT_Info_struct {
-    // the PC of the first inst in this FT
-    Addr start;
-    // the FT size in bytes, counting from the first byte of the first inst to the last byte of the last inst
-    Addr length;
-    // the termination of the FT
-    FT_Ended_By ended_by;
-  } FT_Info;
-
   // Simulator API
   void alloc_mem_decoupled_fe(uns numProcs);
   void init_decoupled_fe(uns proc_id, const char*);
diff --git a/src/ft_info.h b/src/ft_info.h
new file mode 100644
index 0000000..fd8f0b0
--- /dev/null
+++ b/src/ft_info.h
@@ -0,0 +1,62 @@
+/* Copyright 2024 Litz Lab
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/***************************************************************************************
+ * File         : ft_info.h
+ * Author       : Mingsheng Xu <mxu61@ucsc.edu>
+ * Date         : 06/16/2024
+ * Description  :
+ ***************************************************************************************/
+
+#ifndef __FT_INFO_H__
+#define __FT_INFO_H__
+
+typedef enum FT_Ended_By_enum {
+  FT_NOT_ENDED,
+  FT_ICACHE_LINE_BOUNDARY,
+  FT_TAKEN_BRANCH,
+  FT_BAR_FETCH,
+  FT_APP_EXIT
+} FT_Ended_By;
+
+typedef struct FT_Info_Static_struct {
+  // the PC of the first inst in this FT
+  Addr start;
+  // the FT size in bytes, counting from the first byte of the first inst to the last byte of the last inst
+  Addr length;
+  // the number of uops in this FT
+  int n_uops;
+} FT_Info_Static;
+
+// two fts with the same static info may have different dynamic info
+typedef struct FT_Info_Dynamic_struct {
+  // the termination of the FT
+  FT_Ended_By ended_by;
+  // if the first op of this FT is off-path
+  Flag first_op_off_path;
+} FT_Info_Dynamic;
+
+struct FT_Info_struct {
+  FT_Info_Static static_info;
+  FT_Info_Dynamic dynamic_info;
+};
+
+#endif
\ No newline at end of file
diff --git a/src/globals/global_types.h b/src/globals/global_types.h
index d56c09f..9f9e5a2 100644
--- a/src/globals/global_types.h
+++ b/src/globals/global_types.h
@@ -47,6 +47,7 @@ typedef struct Stream_Buffer_struct Stream_Buffer;
 typedef struct Table_Info_struct    Table_Info;
 typedef struct HWP_struct           HWP;
 typedef struct HWP_Info_struct      HWP_Info;
+typedef struct FT_Info_struct       FT_Info;
 
 /* Renames -- Try to use these rather than built-in C types in order to preserve
  * portability */
diff --git a/src/icache_stage.c b/src/icache_stage.c
index b6eda8d..e87c73e 100644
--- a/src/icache_stage.c
+++ b/src/icache_stage.c
@@ -220,6 +220,8 @@ void recover_icache_stage() {
     }
   }
   op_count[ic->proc_id] = bp_recovery_info->recovery_op_num + 1;
+
+  uop_cache_clear_lookup_buffer();
 }
 
 
@@ -312,7 +314,7 @@ void prefetcher_update_on_icache_access(Flag icache_hit) {
 void icache_hit_events(Flag uop_cache_hit) {
   DEBUG(ic->proc_id, "Cache hit on op_num:%s @ 0x%s line_addr 0x%s\n",
         unsstr64(op_count[ic->proc_id]), hexstr64s(ic->fetch_addr), hexstr64s(ic->fetch_addr & ~0x3F));
-  
+
   if (ALWAYS_LOOKUP_ICACHE) {
     if (!ic->off_path) {
       STAT_EVENT(ic->proc_id, ICACHE_HIT_UOP_CACHE_MISS_ON_PATH + uop_cache_hit);
@@ -468,6 +470,9 @@ void update_icache_stage() {
               || ic->state == ICACHE_FINISHED_FT
               || ic->state == ICACHE_FINISHED_FT_EXPECTING_NEXT
               || ic->state == UOP_CACHE_FINISHED_FT) {
+      if (ic->state != ICACHE_FINISHED_FT_EXPECTING_NEXT) {
+        ASSERT(ic->proc_id, !ic->sd.op_count);
+      }
       ASSERT(ic->proc_id, !decoupled_fe_current_ft_can_fetch_op(ic->proc_id));
       if (!decoupled_fe_can_fetch_ft(ic->proc_id)) {
         // if this happened, the app exit should have been seen
@@ -478,11 +483,11 @@ void update_icache_stage() {
         ic->current_ft_info = decoupled_fe_fetch_ft(ic->proc_id);
 
         // set the current fetch address
-        ic->fetch_addr = ic->current_ft_info.start;
+        ic->fetch_addr = ic->current_ft_info.static_info.start;
         ASSERT_PROC_ID_IN_ADDR(ic->proc_id, ic->fetch_addr);
 
         // look up uop cache
-        Flag ft_in_uop_cache = (uop_cache_lookup_line(ic->fetch_addr, ic->current_ft_info.start, ic->current_ft_info.length, FALSE) != NULL);
+        Flag ft_in_uop_cache = uop_cache_lookup_ft_and_fill_lookup_buffer(ic->current_ft_info, ic->off_path);
 
         // look up icache if uop miss (inlcuding when uop cache disabled) or if requested
         if (!ft_in_uop_cache || ALWAYS_LOOKUP_ICACHE) {
@@ -584,20 +589,21 @@ void update_icache_stage() {
         ic->next_state = ICACHE_RETRY_MEM_REQ;
       }
     } else if (ic->state == UOP_CACHE_SERVING) {
-      Uop_Cache_Line_Data* uop_cache_line = uop_cache_lookup_line(ic->fetch_addr, ic->current_ft_info.start, ic->current_ft_info.length, TRUE);
-      ASSERT(ic->proc_id, uop_cache_line);
+      Uop_Cache_Data* uop_cache_line = uop_cache_get_line_from_lookup_buffer();
+      // the line must be valid
+      ASSERT(ic->proc_id, uop_cache_line->n_uops);
 
       ASSERT(ic->proc_id, !ic->sd.op_count);
-      ASSERT(ic->proc_id, uop_cache_line->num_uops <= ISSUE_WIDTH);
-      Flag ft_has_ended = decoupled_fe_fill_icache_stage_data(ic->proc_id, uop_cache_line->num_uops, &ic->sd);
-      ASSERT(ic->proc_id, ic->sd.op_count && ic->sd.op_count == uop_cache_line->num_uops);
+      ASSERT(ic->proc_id, uop_cache_line->n_uops <= ISSUE_WIDTH);
+      Flag ft_has_ended = decoupled_fe_fill_icache_stage_data(ic->proc_id, uop_cache_line->n_uops, &ic->sd);
+      ASSERT(ic->proc_id, ic->sd.op_count && ic->sd.op_count == uop_cache_line->n_uops);
 
       if (ft_has_ended) {
         ASSERT(ic->proc_id, !decoupled_fe_current_ft_can_fetch_op(ic->proc_id));
         // sanity check that the uop cache is in sync
-        ASSERT(ic->proc_id, uop_cache_line->end);
+        ASSERT(ic->proc_id, uop_cache_line->end_of_ft);
         ic->next_state = UOP_CACHE_FINISHED_FT;
-        switch(ic->current_ft_info.ended_by) {
+        switch(ic->current_ft_info.dynamic_info.ended_by) {
           case FT_ICACHE_LINE_BOUNDARY:
           case FT_TAKEN_BRANCH:
             if (ic->sd.op_count < ISSUE_WIDTH) {
@@ -618,9 +624,10 @@ void update_icache_stage() {
           default:
             ASSERT(ic->proc_id, 0);
         }
+        uop_cache_clear_lookup_buffer();
       } else {
         ASSERT(ic->proc_id, decoupled_fe_current_ft_can_fetch_op(ic->proc_id));
-        ASSERT(ic->proc_id, !uop_cache_line->end);
+        ASSERT(ic->proc_id, !uop_cache_line->end_of_ft);
         // the current uop cache assumes that uop cache line op num equals ISSUE_WIDTH
         ASSERT(ic->proc_id, ic->sd.op_count == ISSUE_WIDTH);
         // next_fetch_addr is usually updated when an FT is fetched.
@@ -633,7 +640,7 @@ void update_icache_stage() {
       }
 
       // mark all ops as fetched from the uop cache
-      for(uns ii = 0; ii < STAGE_MAX_OP_COUNT; ii++) {
+      for(int ii = 0; ii < ic->sd.op_count; ii++) {
         ic->sd.ops[ii]->fetched_from_uop_cache = TRUE;
       }
     } else if (ic->state == ICACHE_LOOKUP_SERVING
@@ -655,7 +662,7 @@ void update_icache_stage() {
       if (ft_has_ended) {
         ASSERT(ic->proc_id, !decoupled_fe_current_ft_can_fetch_op(ic->proc_id));
         ic->next_state = ICACHE_FINISHED_FT;
-        switch(ic->current_ft_info.ended_by) {
+        switch(ic->current_ft_info.dynamic_info.ended_by) {
           case FT_ICACHE_LINE_BOUNDARY:
           case FT_TAKEN_BRANCH:
             // if there is more op slots
@@ -753,9 +760,13 @@ void update_icache_stage() {
 
 void update_stats_bf_retired(void) {
   ASSERT(ic->proc_id, !decoupled_fe_current_ft_can_fetch_op(ic->proc_id));
-  FT_Info next_ft_info = decoupled_fe_peek_ft(ic->proc_id);
-  Flag next_op_in_uop_cache = in_uop_cache(next_ft_info.start, FALSE, ic->off_path);
-  if (!next_op_in_uop_cache) {
+  Flag next_ft_in_uop_cache = FALSE;
+  if (decoupled_fe_can_fetch_ft(ic->proc_id)) {
+    FT_Info next_ft_info = decoupled_fe_peek_ft(ic->proc_id);
+    next_ft_in_uop_cache =
+      uop_cache_lookup_line(next_ft_info.static_info.start, next_ft_info, FALSE) != NULL;
+  }
+  if (!next_ft_in_uop_cache) {
     // The micro-op cache can reduce the time to refill the pipeline after a fetch barrier.
     int uop_queue_length = get_uop_queue_stage_length();
     int decode_stages_filled = get_decode_stages_filled();
@@ -784,9 +795,9 @@ static inline void icache_process_ops(void) {
             "Inconsistent off-path op PC: %llx ic:%i op:%i\n", op->inst_info->addr, ic->off_path, op->off_path);
 
     if (!op->off_path) {
-      STAT_EVENT(ic->proc_id, INST_SERVED_BY_ICACHE_ON_PATH + op->fetched_from_uop_cache);
+      STAT_EVENT(ic->proc_id, UOPS_SERVED_BY_ICACHE_ON_PATH + op->fetched_from_uop_cache);
     } else {
-      STAT_EVENT(ic->proc_id, INST_SERVED_BY_ICACHE_OFF_PATH + op->fetched_from_uop_cache);
+      STAT_EVENT(ic->proc_id, UOPS_SERVED_BY_ICACHE_OFF_PATH + op->fetched_from_uop_cache);
     }
 
     if(!op->off_path &&
@@ -1367,12 +1378,6 @@ Flag instr_fill_line(Mem_Req* req) {
     icache_fill_line(req);
     DEBUG(ic->proc_id, "line 0x%llx is filled into icache\n", req->addr);
   }
-  // TEMP CHANGE: all FDIPPRF are UOC_PREF as well, except for a corner case where branch target is same line
-  if (UOC_PREF && (mem_req_is_type(req, MRT_FDIPPRFON) || mem_req_is_type(req, MRT_FDIPPRFOFF) || mem_req_is_type(req, MRT_UOCPRF))) {
-    if (in_uop_cache(req->addr, FALSE, req->off_path)) {
-      STAT_EVENT(ic->proc_id, UOP_CACHE_HIT_NO_PREFETCH);
-    }
-  }
 
   return TRUE;
 }
diff --git a/src/libs/cpp_cache.cc b/src/libs/cpp_cache.cc
deleted file mode 100644
index c93149b..0000000
--- a/src/libs/cpp_cache.cc
+++ /dev/null
@@ -1,231 +0,0 @@
-// CPP implementation of a cache.
-// Built-in support for multi-line units, e.g. PWs that take multiple lines.
-
-#include "libs/cpp_cache.h"
-#include <cstdlib>
-#include <iostream>
-#include <list>
-#include <string>
-#include <unordered_map>
-#include <vector>
-#include "debug/debug_macros.h"
-#include "debug/debug_print.h"
-
-extern "C" {
-#include "globals/assert.h"
-#include "globals/utils.h"
-#include "memory/memory.param.h"
-}
-
-#define DEBUG(proc_id, args...) _DEBUG(proc_id, DEBUG_CPP_CACHE, ##args)
-
-class Entry {
- public:
-  Addr    addr;
-  Addr    tag;
-  int     num_lines;
-  bool    priority;
-  void*   data;
-};
-
-// List with custom capacity logging, enabling multiline Entries.
-class Set {
- public:
-  std::list<Entry> lst       = {};
-  size_t           num_lines = 0;
-};
-
-class Cpp_Cache {
- private:
-  std::vector<Set> sets;
-  Repl_Policy      repl_policy;
-  Counter          capacity_lines;
-  uns              assoc;
-  uns              line_bytes;
-  bool             tag_incl_offset;
-  int              data_size;
-
-  uns offset_bits;
-  uns offset_mask;
-  uns set_mask;
-  uns tag_mask;
-  uns line_addr_mask;
-
- public:
-  Cpp_Cache(uns nl, uns asc, uns lb, Repl_Policy rp, bool tio, int ds) {
-    capacity_lines  = nl;
-    assoc           = asc;
-    line_bytes      = lb;
-    repl_policy     = rp;
-    tag_incl_offset = tio;
-    data_size       = ds;
-
-    uns num_sets = capacity_lines / assoc;
-    sets         = std::vector<Set>(num_sets);
-    offset_bits  = LOG2(line_bytes);
-    offset_mask  = N_BIT_MASK(offset_bits);
-    set_mask     = N_BIT_MASK(LOG2(num_sets)) << offset_bits;
-    // If tag_incl_offset, we want to use non-LSB bits for set although cache is
-    // byte-addressable
-    tag_mask       = tag_incl_offset ? ~set_mask : ~(set_mask & offset_mask);
-    line_addr_mask = tag_incl_offset ? N_BIT_MASK_64 : ~offset_mask;
-  }
-
-  uns   index(Addr addr, Addr* tag, Addr* line_addr);
-  void* access(Addr addr, bool update_repl, bool upgrade_priority);
-  void* insert(Addr addr, int lines_used, bool priority, Addr* evicted_entries);
-  Addr  evict(Set& set);
-  void  insert_entry_at_pos(Set& set, Entry entry,
-                            std::list<Entry>::iterator pos);
-};
-
-std::unordered_map<std::string, Cpp_Cache> cache_map{};
-
-// utility function for debugging
-std::ostream& operator<<(std::ostream& os, const std::list<Entry>& list) {
-  for(auto const& entry : list) {
-    os << entry.tag << ", num_lines=" << entry.num_lines << "; ";
-  }
-  os << std::endl;
-  return os;
-}
-
-
-static inline Cpp_Cache* get_cache(char* name) {
-  std::string key = name;
-  return &cache_map.at(key);
-}
-
-// returns the set index of the addr within this cache.
-uns cpp_cache_index(char* name, Addr addr, Addr* tag, Addr* line_addr) {
-  return get_cache(name)->index(addr, tag, line_addr);
-}
-
-void cpp_cache_create(char* name, uns num_lines, uns assoc, uns line_bytes,
-                      Repl_Policy repl_policy, Flag tag_incl_offset,
-                      int data_size) {
-  Cpp_Cache   cache(num_lines, assoc, line_bytes, repl_policy,
-                    tag_incl_offset == TRUE, data_size);
-  std::string key = name;
-  cache_map.insert({key, cache});
-}
-
-// Insert into cache, returning a void* for the callee to fill with data.
-// evicted_entries must have enough allocated space to capture num_addr up to the cache assoc.
-void* cpp_cache_insert(char* name, Addr addr, int lines_used, Flag priority, Addr* evicted_entries) {
-  return get_cache(name)->insert(addr, lines_used, priority == TRUE, evicted_entries); //notify which lines evicted? Could be multiple! require an array to be passed in, size of assoc.
-}
-
-// Access cache. Upgrade priority if upgrade_priority and update_repl.
-void* cpp_cache_access(char* name, Addr addr, Flag update_repl, Flag upgrade_priority) {
-  return get_cache(name)->access(addr, update_repl == TRUE, upgrade_priority == TRUE);;
-}
-
-/////////// Cache Member Methods ////////////
-
-uns Cpp_Cache::index(Addr addr, Addr* tag, Addr* line_addr) {
-  // may need to return tag or line_addr as well.
-  *tag       = addr & tag_mask;
-  *line_addr = addr & line_addr_mask;
-  return (addr & set_mask) >> offset_bits;
-}
-
-// access: Looks up the cache based on addr. Returns pointer to line data if
-// found.
-void* Cpp_Cache::access(Addr addr, bool update_repl, bool upgrade_priority) {
-  Addr  tag;
-  Addr  line_addr;
-  void* data    = NULL;
-  uns   set_idx = index(addr, &tag, &line_addr);
-  for(std::list<Entry>::iterator entry_it = sets.at(set_idx).lst.begin();
-      entry_it != sets.at(set_idx).lst.end(); ++entry_it) {
-    Entry entry = *entry_it;
-    if(entry.tag == tag) {  // hit
-      data = entry.data;
-      if(update_repl) {
-        if (upgrade_priority) {
-          entry.priority = true;
-        }
-        // Move to LRU position: remove entry from list and reinsert at Head
-        sets.at(set_idx).lst.erase(entry_it);
-        insert_entry_at_pos(sets.at(set_idx), entry,
-                            sets.at(set_idx).lst.begin());
-      }
-      break;
-    }
-  }
-  return data;
-}
-
-// Returns evicted line addr
-Addr Cpp_Cache::evict(Set& set) {
-  Addr evicted_line;
-  ASSERT(0, !set.lst.empty());
-  auto it_evicted_entry = --set.lst.end();  // last element
-  set.num_lines -= it_evicted_entry->num_lines;
-  evicted_line = it_evicted_entry->addr;
-  free(it_evicted_entry->data);
-  set.lst.erase(it_evicted_entry);
-  return evicted_line;
-}
-
-// there are potentially Different policies for insert and access.
-// A different insert policy would choose a different insert pos
-// OR for insert, first insert at end, then update_repl for it.
-void Cpp_Cache::insert_entry_at_pos(Set& set, Entry entry_to_insert,
-                                    std::list<Entry>::iterator pos) {
-  switch(repl_policy) {
-    case REPL_TRUE_LRU:
-      set.lst.insert(pos, entry_to_insert);
-      break;
-    case REPL_STICKY_PRIORITY_LINES: {
-      if(rand() % 100 > PRIORITY_LINE_STICKINESS_PERCENT) {  // if not sticky
-        set.lst.insert(pos, entry_to_insert);
-        break;
-      }
-      // Find insert point.
-      std::list<Entry>::iterator insert_pos = pos;
-      while(insert_pos != set.lst.end() && !entry_to_insert.priority && insert_pos->priority) {  //what if all are prio?  *************
-        ++insert_pos;
-        // now insert_pos is at the element right after the insert position
-      }
-      set.lst.insert(insert_pos, entry_to_insert);
-      for(auto cur_entry_it = insert_pos; cur_entry_it != set.lst.end();
-          ++cur_entry_it) {
-        auto prev_entry_it = std::prev(cur_entry_it);
-        if(cur_entry_it->priority && !prev_entry_it->priority) {
-          // move cur_entry left one element in the list
-          set.lst.splice(prev_entry_it, set.lst, cur_entry_it);
-        }
-      }
-    } break;
-    default:
-      ASSERT(0, FALSE);  // unsupported
-  }
-}
-
-void* Cpp_Cache::insert(Addr addr, int lines_used, bool priority, Addr* evicted_entries) {
-  Addr  tag;
-  Addr  line_addr;
-  void* data    = NULL;
-  uns   set_idx = index(addr, &tag, &line_addr);
-  // First check if line exists
-  for(Entry entry : sets.at(set_idx).lst) {
-    if(entry.tag == tag) {  // hit, do nothing
-      return NULL;
-    }
-  }
-  data = calloc(1, data_size);
-  Entry new_entry{addr, tag, lines_used, priority, data};
-  // Evict any entries required
-  uns num_entries_evicted = 0;
-  while(assoc < sets.at(set_idx).num_lines + new_entry.num_lines) {
-    evicted_entries[num_entries_evicted++] = evict(sets.at(set_idx));
-  }
-  ASSERT(0, num_entries_evicted <= assoc);
-  insert_entry_at_pos(sets.at(set_idx), new_entry,
-                      sets.at(set_idx).lst.begin());
-  sets.at(set_idx).num_lines += lines_used;
-  ASSERT(0, sets.at(set_idx).num_lines <= assoc);
-  return data;
-}
diff --git a/src/libs/cpp_cache.h b/src/libs/cpp_cache.h
index 09ea979..e6a68b8 100644
--- a/src/libs/cpp_cache.h
+++ b/src/libs/cpp_cache.h
@@ -1,26 +1,11 @@
 // CPP implementation of a cache.
-// Built-in support for multi-line units, e.g. PWs that take multiple lines.
 
 #ifndef __CPP_CACHE_H__
 #define __CPP_CACHE_H__
 
-#ifdef __cplusplus
-extern "C" {
-#endif
-
 #include "globals/global_types.h"
 #include "cache_lib.h"
 
-void cpp_cache_create(char* name, uns num_lines, uns assoc, uns line_bytes,
-                      Repl_Policy repl_policy, Flag tag_incl_offset, int data_size);
-uns cpp_cache_index(char* name, Addr addr, Addr* tag, Addr* line_addr);
-void* cpp_cache_insert(char* name, Addr addr, int lines_used, Flag priority, Addr* evicted_entries);
-void* cpp_cache_access(char* name, Addr addr, Flag update_repl, Flag upgrade_priority);
-
-
-
-#ifdef __cplusplus
-}
-#endif
+#include "cpp_cache.tpp"
 
 #endif
\ No newline at end of file
diff --git a/src/libs/cpp_cache.tpp b/src/libs/cpp_cache.tpp
new file mode 100644
index 0000000..552730e
--- /dev/null
+++ b/src/libs/cpp_cache.tpp
@@ -0,0 +1,173 @@
+// CPP implementation of a cache.
+
+#include "libs/cpp_cache.h"
+#include <iostream>
+#include <list>
+#include <unordered_map>
+#include <vector>
+#include "debug/debug_macros.h"
+#include "debug/debug_print.h"
+
+extern "C" {
+#include "globals/assert.h"
+#include "globals/utils.h"
+#include "globals/global_vars.h"
+#include "memory/memory.param.h"
+}
+
+#define CPPC_DEBUG(proc_id, args...) _DEBUG(proc_id, DEBUG_CPP_CACHE, ##args)
+
+template <typename User_Key_Type, typename User_Data_Type>
+struct Entry {
+  Flag valid;
+  User_Key_Type key;
+  User_Data_Type data;
+  // for LRU replacement policy
+  Counter accessed_cycle;
+};
+
+template <typename User_Key_Type, typename User_Data_Type>
+class Set {
+ public:
+  std::vector<Entry<User_Key_Type, User_Data_Type>> entries;
+  // for round-robin replacement policy
+  uns next_evict;
+};
+
+template <typename User_Key_Type, typename User_Data_Type>
+class Cpp_Cache {
+ protected:
+  std::vector<Set<User_Key_Type, User_Data_Type>> sets;
+  Repl_Policy      repl_policy;
+  uns              assoc;
+  uns              num_sets;
+  uns              line_bytes;
+
+  // defines how to hash the key to the set index, need to be implemented by the user
+  virtual uns set_idx_hash(User_Key_Type key) = 0;
+
+  // replacement policy functions
+  virtual void update_repl_states(Set<User_Key_Type, User_Data_Type>& set, uns hit_idx);
+  virtual uns get_repl_idx(Set<User_Key_Type, User_Data_Type>& set);
+
+ public:
+  Cpp_Cache() = default;
+  Cpp_Cache(uns nl, uns asc, uns lb, Repl_Policy rp) {
+    assoc           = asc;
+    num_sets        = nl / assoc;
+    line_bytes      = lb;
+    repl_policy     = rp;
+
+    sets            = std::vector<Set<User_Key_Type, User_Data_Type>>(num_sets);
+    for (uns i = 0; i < num_sets; i++) {
+      sets[i].entries.resize(assoc);
+    }
+  }
+
+  User_Data_Type* access(User_Key_Type key, bool update_repl);
+  Entry<User_Key_Type, User_Data_Type> insert(User_Key_Type key, User_Data_Type data);
+  Entry<User_Key_Type, User_Data_Type> invalidate(User_Key_Type key);
+};
+
+template <typename User_Key_Type, typename User_Data_Type>
+void Cpp_Cache<User_Key_Type, User_Data_Type>::update_repl_states(Set<User_Key_Type, User_Data_Type>& set, uns hit_idx) {
+  switch(repl_policy) {
+    case REPL_TRUE_LRU: {
+      set.entries[hit_idx].accessed_cycle = cycle_count;
+    } break;
+    case REPL_RANDOM: {
+    } break;
+    case REPL_ROUND_ROBIN: {
+    } break;
+    default:
+      ASSERT(0, FALSE);  // unsupported
+  }
+}
+
+template <typename User_Key_Type, typename User_Data_Type>
+uns Cpp_Cache<User_Key_Type, User_Data_Type>::get_repl_idx(Set<User_Key_Type, User_Data_Type>& set) {
+  // if there are invalid entries, replace them
+  for (uns i = 0; i < assoc; i++) {
+    Entry<User_Key_Type, User_Data_Type> entry = set.entries[i];
+    if (!entry.valid) {
+      return i;
+    }
+  }
+
+  // otherwise, replace according to the policy
+  uns repl_idx = 0;
+  switch(repl_policy) {
+    case REPL_TRUE_LRU: {
+      Counter lru_cycle = std::numeric_limits<Counter>::max();
+      for (uns i = 0; i < assoc; i++) {
+        Entry<User_Key_Type, User_Data_Type> entry = set.entries[i];
+        // find smallest access cycle
+        if (entry.accessed_cycle < lru_cycle) {
+          repl_idx = i;
+          lru_cycle = entry.accessed_cycle;
+        }
+      }
+    } break;
+    case REPL_RANDOM: {
+      repl_idx = rand() % assoc;
+    } break;
+    case REPL_ROUND_ROBIN: {
+      repl_idx = (set.next_evict + 1) % assoc;
+      set.next_evict = repl_idx;
+    } break;
+    default:
+      ASSERT(0, FALSE);  // unsupported
+  }
+  return repl_idx;
+}
+
+// access: Looks up the cache based on key. Returns pointer to line data if found
+template <typename User_Key_Type, typename User_Data_Type>
+User_Data_Type* Cpp_Cache<User_Key_Type, User_Data_Type>::access(User_Key_Type key, bool update_repl) {
+  User_Data_Type* data = NULL;
+  uns set_idx = set_idx_hash(key);
+  for (uns i = 0; i < assoc; i++) {
+    Entry<User_Key_Type, User_Data_Type>& entry = sets[set_idx].entries[i];
+    if (entry.valid && entry.key == key) {  // hit
+      data = &entry.data;
+      if (update_repl) {
+        update_repl_states(sets[set_idx], i);
+      }
+      break;
+    }
+  }
+  return data;
+}
+
+template <typename User_Key_Type, typename User_Data_Type>
+Entry<User_Key_Type, User_Data_Type> Cpp_Cache<User_Key_Type, User_Data_Type>::insert(User_Key_Type key, User_Data_Type data) {
+  // first check if line exists
+  ASSERT(0, access(key, FALSE) == NULL);
+
+  uns set_idx = set_idx_hash(key);
+  // if the set is full, repl_idx will be overwritten;
+  // if the set has vacancy, repl_idx will point to an invalid line
+  uns repl_idx = get_repl_idx(sets[set_idx]);
+
+  Entry<User_Key_Type, User_Data_Type> evicted_entry = sets[set_idx].entries[repl_idx];
+  sets[set_idx].entries[repl_idx] = Entry<User_Key_Type, User_Data_Type>{TRUE, key, data, 0};
+  update_repl_states(sets[set_idx], repl_idx);
+
+  // the evicted entry is an eviciton victim if and only if it is valid
+  return evicted_entry;
+}
+
+template <typename User_Key_Type, typename User_Data_Type>
+Entry<User_Key_Type, User_Data_Type> Cpp_Cache<User_Key_Type, User_Data_Type>::invalidate(User_Key_Type key) {
+  uns set_idx = set_idx_hash(key);
+  Entry<User_Key_Type, User_Data_Type> invalidated_entry{};
+  for (uns i = 0; i < assoc; i++) {
+    Entry<User_Key_Type, User_Data_Type>& entry = sets[set_idx].entries[i];
+    if (entry.valid && entry.key == key) {  // hit
+      invalidated_entry = entry;
+      entry.valid = FALSE;
+      break;
+    }
+  }
+  return invalidated_entry;
+}
\ No newline at end of file
diff --git a/src/memory/memory.param.def b/src/memory/memory.param.def
index a218386..df8f9be 100644
--- a/src/memory/memory.param.def
+++ b/src/memory/memory.param.def
@@ -114,9 +114,6 @@ DEF_PARAM(icache_repl, ICACHE_REPL, uns, uns, 0, )
 /* Uop Cache */
 DEF_PARAM(uop_cache_enable, UOP_CACHE_ENABLE, uns, uns, 1,)
 DEF_PARAM(always_lookup_icache, ALWAYS_LOOKUP_ICACHE, uns, uns, 0,)
-DEF_PARAM(  oracle_perfect_uop_cache       , ORACLE_PERFECT_UOP_CACHE        , Flag   , Flag    , FALSE                  ,       )
-DEF_PARAM(  inf_size_uop_cache             , INF_SIZE_UOP_CACHE              , Flag   , Flag    , FALSE                  ,       )
-DEF_PARAM(  inf_size_uop_cache_pw_size_lim , INF_SIZE_UOP_CACHE_PW_SIZE_LIM  , uns    , uns     , 0                      ,       )
 DEF_PARAM(uop_cache_insert_only_onpath, UOP_CACHE_INSERT_ONLY_ONPATH, Flag, Flag, FALSE,)
 // Uop cache size in its uop capacity = ISSUE_WIDTH * UOP_CACHE_LINES = 6 * 384 = 2304 uops
 DEF_PARAM(  uop_cache_lines                , UOP_CACHE_LINES                 , uns    , uns     , 384                    ,       )
@@ -131,9 +128,6 @@ DEF_PARAM(priority_line_stickiness_percent, PRIORITY_LINE_STICKINESS_PERCENT, in
 DEF_PARAM(pw_priority_list_filepath, PW_PRIORITY_LIST_FILEPATH, char*, string, "pw_priority_list.csv", )
 DEF_PARAM(num_priority_pws, NUM_PRIORITY_PWS, uns, uns, 10,)
 DEF_PARAM(prioritize_pws_after_first_sticky_until_backend_stall, PRIORITIZE_PWS_AFTER_FIRST_STICKY_UNTIL_BACKEND_STALL, Flag, Flag, FALSE,)
-// If set, PWs are only inserted after a resteer until the uop queue fills.
-// The UOC becomes a cache only for ops that follow a resteer.
-DEF_PARAM(uop_cache_insert_only_after_resteer_uop_queue_not_full, UOP_CACHE_INSERT_ONLY_AFTER_RESTEER_UOP_QUEUE_NOT_FULL, Flag, Flag, FALSE,)
 DEF_PARAM(no_repl_after_queue_size, NO_REPL_AFTER_QUEUE_SIZE, uns, uns, 0,)
 /* FDIP Confidence Cache */
 // The cache line size is fixed to ICACHE_LINE_SIZE (64 by default). The number of entries in a FDIP UC is FDIP_UC_SIZE / ICACHE_LINE_SIZE. To have 16K (16384) entries, FDIP_UC_SIZE should be 1048576.
diff --git a/src/memory/memory.stat.def b/src/memory/memory.stat.def
index 0d624f7..718773a 100644
--- a/src/memory/memory.stat.def
+++ b/src/memory/memory.stat.def
@@ -107,11 +107,11 @@ DEF_STAT(FT_UOP_CACHE_HIT_OFF_PATH, DIST, NO_RATIO)
 DEF_STAT(FT_UOP_CACHE_MISS_ICACHE_HIT_OFF_PATH, COUNT, NO_RATIO)
 DEF_STAT(FT_UOP_CACHE_MISS_ICACHE_MISS_OFF_PATH, DIST, NO_RATIO)
 
-DEF_STAT(INST_SERVED_BY_ICACHE_ON_PATH, DIST, NO_RATIO)
-DEF_STAT(INST_SERVED_BY_UOP_CACHE_ON_PATH, DIST, NO_RATIO)
+DEF_STAT(UOPS_SERVED_BY_ICACHE_ON_PATH, DIST, NO_RATIO)
+DEF_STAT(UOPS_SERVED_BY_UOP_CACHE_ON_PATH, DIST, NO_RATIO)
 
-DEF_STAT(INST_SERVED_BY_ICACHE_OFF_PATH, DIST, NO_RATIO)
-DEF_STAT(INST_SERVED_BY_UOP_CACHE_OFF_PATH, DIST, NO_RATIO)
+DEF_STAT(UOPS_SERVED_BY_ICACHE_OFF_PATH, DIST, NO_RATIO)
+DEF_STAT(UOPS_SERVED_BY_UOP_CACHE_OFF_PATH, DIST, NO_RATIO)
 
 // Micro Op Cache
 DEF_STAT(ICACHE_HIT_UOP_CACHE_MISS_ON_PATH, DIST, NO_RATIO)
@@ -124,32 +124,35 @@ DEF_STAT(ICACHE_HIT_UOP_CACHE_HIT_OFF_PATH, COUNT, NO_RATIO)
 DEF_STAT(ICACHE_MISS_UOP_CACHE_MISS_OFF_PATH, COUNT, NO_RATIO)
 DEF_STAT(ICACHE_MISS_UOP_CACHE_HIT_OFF_PATH, DIST, NO_RATIO)
 
-DEF_STAT(UOP_CACHE_MISS, DIST, NO_RATIO)
-DEF_STAT(UOP_CACHE_HIT, DIST, NO_RATIO)
+DEF_STAT(UOP_CACHE_FT_INSERTED_ONPATH_USED_ONPATH, COUNT, NO_RATIO)
+DEF_STAT(UOP_CACHE_FT_INSERTED_OFFPATH_USED_ONPATH, COUNT, NO_RATIO)
 
-DEF_STAT(UOP_CACHE_MISS_ONPATH, DIST, NO_RATIO)
-DEF_STAT(UOP_CACHE_MISS_OFFPATH, DIST, NO_RATIO)
+DEF_STAT(UOP_CACHE_LINE_INSERTED_ONPATH_USED_ONPATH, COUNT, NO_RATIO)
+DEF_STAT(UOP_CACHE_LINE_INSERTED_OFFPATH_USED_ONPATH, COUNT, NO_RATIO)  // (off path doesnt matter?)
 
-DEF_STAT(UOP_CACHE_HIT_ONPATH, DIST, NO_RATIO)
-DEF_STAT(UOP_CACHE_HIT_OFFPATH, DIST, NO_RATIO)
+DEF_STAT(  UOP_CACHE_PREFETCH           , COUNT   , NO_RATIO)
 
-DEF_STAT(UOP_CACHE_PWS_INSERTED_ONPATH, DIST, NO_RATIO)
-DEF_STAT(UOP_CACHE_PWS_INSERTED_OFFPATH, DIST, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_INSERT_FAILED_INST_TOO_BIG_ON_PATH   , DIST, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_INSERT_FAILED_FT_TOO_BIG_ON_PATH     , COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_INSERT_CONFLICTED_SHORT_REUSE_ON_PATH, COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_INSERT_SUCCEEDED_ON_PATH             , DIST, NO_RATIO)
 
-DEF_STAT(UOP_CACHE_LINES_INSERTED_ONPATH, DIST, NO_RATIO)
-DEF_STAT(UOP_CACHE_LINES_INSERTED_OFFPATH, DIST, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_INSERT_FAILED_INST_TOO_BIG_OFF_PATH   , DIST, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_INSERT_FAILED_FT_TOO_BIG_OFF_PATH     , COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_INSERT_CONFLICTED_SHORT_REUSE_OFF_PATH, COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_INSERT_SUCCEEDED_OFF_PATH             , DIST, NO_RATIO)
 
-DEF_STAT(UOP_CACHE_LINES_INSERTED_ONPATH_USED_ONPATH, COUNT, NO_RATIO)
-DEF_STAT(UOP_CACHE_LINES_INSERTED_OFFPATH_USED_ONPATH, COUNT, NO_RATIO)  // (off path doesnt matter?)
+DEF_STAT(  UOP_CACHE_LINE_INSERT_FAILED_INST_TOO_BIG_ON_PATH   , DIST, NO_RATIO)
+DEF_STAT(  UOP_CACHE_LINE_INSERT_FAILED_FT_TOO_BIG_ON_PATH     , COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_LINE_INSERT_CONFLICTED_SHORT_REUSE_ON_PATH, COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_LINE_INSERT_SUCCEEDED_ON_PATH             , DIST, NO_RATIO)
 
-DEF_STAT(  UOP_CACHE_PREFETCH           , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_HIT_NO_PREFETCH    , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_INSERT_FAILED_OFFPATH, COUNT, NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_INSERT_FAILED_TOO_LONG    , COUNT, NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_INSERT_FAILED_CACHE_HIT   , COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_LINE_INSERT_FAILED_INST_TOO_BIG_OFF_PATH   , DIST, NO_RATIO)
+DEF_STAT(  UOP_CACHE_LINE_INSERT_FAILED_FT_TOO_BIG_OFF_PATH     , COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_LINE_INSERT_CONFLICTED_SHORT_REUSE_OFF_PATH, COUNT, NO_RATIO)
+DEF_STAT(  UOP_CACHE_LINE_INSERT_SUCCEEDED_OFF_PATH             , DIST, NO_RATIO)
 
 DEF_STAT(RECOVER_AT_DECODE_BR_FROM_UOC, COUNT, NO_RATIO)
-DEF_STAT(RECOVERED_FROM_UOC, COUNT, NO_RATIO)
 
 DEF_STAT(  UOP_CACHE_ICACHE_SWITCH_BR_NOT_TAKEN_RESTEERED   , DIST    , NO_RATIO)
 DEF_STAT(  UOP_CACHE_ICACHE_SWITCH_BR_NOT_TAKEN_CORRECT_PRED, COUNT   , NO_RATIO)
@@ -291,31 +294,26 @@ DEF_STAT(  BF_UOP_CACHE_MISS_UOP_QUEUE_PLUS_DECODE_LENGTH_18      , COUNT   , NO
 DEF_STAT(  BF_UOP_CACHE_MISS_UOP_QUEUE_PLUS_DECODE_LENGTH_19      , COUNT   , NO_RATIO)
 DEF_STAT(  BF_UOP_CACHE_MISS_UOP_QUEUE_PLUS_DECODE_LENGTH_20      , DIST    , NO_RATIO)
 
-DEF_STAT(  UOP_CACHE_PW_LENGTH_1             , DIST    , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_2             , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_3             , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_4             , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_5             , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_6             , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_7             , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_8             , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_9             , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_10            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_11            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_12            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_13            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_14            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_15            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_16            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_17            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_18            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_19            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_20            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_21            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_22            , COUNT   , NO_RATIO)
-DEF_STAT(  UOP_CACHE_PW_LENGTH_23            , COUNT   , NO_RATIO)
-DEF_STAT(UOP_CACHE_PW_LENGTH_24, COUNT, NO_RATIO)
-DEF_STAT(UOP_CACHE_PW_LENGTH_MAX, DIST, NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_1_ON_PATH             , DIST    , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_2_ON_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_3_ON_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_4_ON_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_5_ON_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_6_ON_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_7_ON_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_8_ON_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_9_AND_MORE_ON_PATH    , DIST    , NO_RATIO)
+
+DEF_STAT(  UOP_CACHE_FT_LINES_1_OFF_PATH             , DIST    , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_2_OFF_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_3_OFF_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_4_OFF_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_5_OFF_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_6_OFF_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_7_OFF_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_8_OFF_PATH             , COUNT   , NO_RATIO)
+DEF_STAT(  UOP_CACHE_FT_LINES_9_AND_MORE_OFF_PATH    , DIST    , NO_RATIO)
+
 
 DEF_STAT(  UOP_CACHE_PREFETCH_PW_LENGTH_1             , DIST    , NO_RATIO)
 DEF_STAT(  UOP_CACHE_PREFETCH_PW_LENGTH_2             , COUNT   , NO_RATIO)
diff --git a/src/op.h b/src/op.h
index 54a99f8..1474f13 100644
--- a/src/op.h
+++ b/src/op.h
@@ -34,6 +34,7 @@
 #include "inst_info.h"
 #include "op_info.h"
 #include "table_info.h"
+#include "ft_info.h"
 
 
 /**************************************************************************************/
@@ -97,25 +98,6 @@ typedef struct Per_Branch_Stat_struct {
   Addr addr;
   Cf_Type cf_type;
   Addr target;
-  int bpu_hit_uc_hit_on_path;
-  int bpu_hit_uc_miss_on_path;
-  int bpu_hit_uc_ic_miss_on_path;
-  int bpu_hit_uc_hit_off_path;
-  int bpu_hit_uc_miss_off_path;
-  int bpu_hit_uc_ic_miss_off_path;
-  int mispred_uc_hit;
-  int mispred_uc_miss;
-  int mispred_uc_ic_miss;   // mispred that miss in both uc and ic
-  int misfetch_uc_hit;
-  int misfetch_uc_miss;
-  int misfetch_uc_ic_miss;
-  int btb_miss_uc_hit;
-  int btb_miss_uc_miss;
-  int btb_miss_uc_ic_miss;
-  int other_recovery_uc_hit;
-  int other_recovery_uc_miss;
-  int other_recovery_uc_ic_miss;
-  int recover_redirect_extra_fetch_latency; // extra stall cycles due to target not being in UC
 } Per_Branch_Stat;
 
 /*------------------------------------------------------------------------------------*/
@@ -324,6 +306,7 @@ struct Op_struct {
   Flag redirect_scheduled;
   // }}}
 
+  FT_Info ft_info;  // FT the op associated with
   // {{{ uop cache
   Flag fetched_from_uop_cache;
   // }}}
@@ -335,18 +318,6 @@ struct Op_struct {
 };
 // }}}
 
-// Contains the first and last addresses in the prediction window
-typedef struct Uop_Cache_Data_struct {
-  Addr first;
-  Addr last;
-  Counter n_uops;
-  Counter last_op_num;
-  Flag prefetch;
-  Counter used;
-  Flag first_op_offpath;
-  Flag priority;
-} Uop_Cache_Data;
-
 /**************************************************************************************/
 
 #endif  // #ifndef __OP_H__
diff --git a/src/pin/pin_lib/uop_generator.c b/src/pin/pin_lib/uop_generator.c
index 6077e1b..be52433 100644
--- a/src/pin/pin_lib/uop_generator.c
+++ b/src/pin/pin_lib/uop_generator.c
@@ -1004,7 +1004,7 @@ void convert_dyn_uop(uns8 proc_id, Inst_Info* info, ctype_pin_inst* pi,
 
   trace_uop->exit = is_last_uop ? pi->exit : 0;
 
-  trace_uop->npc = trace_uop->addr;  // CHECKME!!
+  trace_uop->npc = trace_uop->info->addr;
 }
 
 void uop_generator_recover(uns8 proc_id) {
diff --git a/src/sim.c b/src/sim.c
index 24c34f5..eb63f98 100644
--- a/src/sim.c
+++ b/src/sim.c
@@ -281,17 +281,10 @@ static inline void check_heartbeat(uns8 proc_id, Flag final) {
                   unsstr64(sim_time), cum_ipc, cum_ipc, cum_khz);
           FILE* fp = fopen("per_branch_stats.csv", "w");
           Per_Branch_Stat** entries = (Per_Branch_Stat**) hash_table_flatten(&per_branch_stat, NULL);
-          fprintf(fp, "cf_type,addr,target,bpu_hit_uc_hit,bpu_hit_uc_miss,mispred_uc_hit,"
-                  "mispred_uc_miss,misfetch_uc_hit,misfetch_uc_miss,btb_miss_uc_hit,"
-                  "btb_miss_uc_miss,other_recovery_uc_hit,other_recovery_uc_miss,"
-                  "recover_redirect_extra_fetch_latency\n");
+          fprintf(fp, "cf_type,addr,target\n");
           for (int i=0; i<per_branch_stat.count; i++) {
             Per_Branch_Stat* entry = entries[i];
-            fprintf(fp, "%i,%llx,%llx,%i,%i,%i,%i,%i,%i,%i,%i,%i,%i,%i\n", entry->cf_type, entry->addr, 
-              entry->target, entry->bpu_hit_uc_hit_on_path, entry->bpu_hit_uc_miss_on_path, entry->mispred_uc_hit,
-              entry->mispred_uc_miss, entry->misfetch_uc_hit, entry->misfetch_uc_miss, 
-              entry->btb_miss_uc_hit, entry->btb_miss_uc_miss, entry->other_recovery_uc_hit,
-              entry->other_recovery_uc_miss, entry->recover_redirect_extra_fetch_latency);
+            fprintf(fp, "%i,%llx,%llx\n", entry->cf_type, entry->addr, entry->target);
           }
           free(entries);
 
diff --git a/src/uop_cache.c b/src/uop_cache.c
deleted file mode 100644
index 3c65a07..0000000
--- a/src/uop_cache.c
+++ /dev/null
@@ -1,345 +0,0 @@
-/***************************************************************************************
- * File         : uop_cache.h
- * Author       : Peter Braun
- * Date         : 10.28.2020
- * Description  : Interface for interacting with uop cache object.
- *                  Following Kotra et. al.'s MICRO 2020 description of uop cache baseline
- ***************************************************************************************/
-
-#include "debug/debug_macros.h"
-#include "debug/debug_print.h"
-#include "globals/assert.h"
-#include "globals/global_defs.h"
-#include "globals/global_types.h"
-#include "globals/global_vars.h"
-#include "globals/utils.h"
-#include "isa/isa_macros.h"
-
-#include "bp/bp.h"
-#include "op_pool.h"
-
-#include "core.param.h"
-#include "debug/debug.param.h"
-#include "general.param.h"
-#include "statistics.h"
-
-#include "libs/cache_lib.h"
-#include "memory/memory.h"
-#include "memory/memory.param.h"
-#include "uop_cache.h"
-#include "icache_stage.h"
-#include "libs/cpp_cache.h"
-#include "uop_queue_stage.h"
-
-/**************************************************************************************/
-/* Macros */
-
-#define DEBUG(proc_id, args...) _DEBUG(proc_id, DEBUG_UOP_CACHE, ##args)
-
-// Uop cache is byte-addressable, so tag/set index are generated from full address (no offset)
-// Uop cache uses icache tag + icache offset as full TAG
-#define UOP_CACHE_LINE_SIZE       ICACHE_LINE_SIZE
-#define UOP_QUEUE_SIZE            1000 // at least UOP_CACHE_ASSOC * ISSUE_WIDTH
-#define UOP_CACHE_LINE_DATA_SIZE  sizeof(Uop_Cache_Data)
-#define UOP_CACHE_NAME            "UOP_CACHE"
-
-/**************************************************************************************/
-/* Local Prototypes */
-
-static inline Flag insert_uop_cache(void);
-static inline Flag in_uop_cache_search(Addr search_addr, Flag update_repl, Flag offpath);
-
-/**************************************************************************************/
-/* Global Variables */
-
-uns8 proc_id;
-static Flag UOP_CACHE_ON;
-
-// uop trace/bbl accumulation
-static Uop_Cache_Data accumulating_pw = {0};
-static Counter next_accum_op = 0;
-static Op* uop_q[UOP_QUEUE_SIZE];
-
-// k: instr addr
-Hash_Table inf_size_uop_cache;
-// indexed by start addr of PW
-Hash_Table pc_to_pw;
-
-Addr addr_following_resteer_bf = 0;  // Addr that follows a resteer or fetch barrier
-Flag uop_cache_insert_enable = TRUE;
-
-/**************************************************************************************/
-/* init_uop_cache */
-
-void init_uop_cache(uns8 pid) {
-  proc_id = pid;
-  UOP_CACHE_ON = UOP_CACHE_ENABLE && (UOP_CACHE_LINES || INF_SIZE_UOP_CACHE || ORACLE_PERFECT_UOP_CACHE);
-  if (!UOP_CACHE_ON) {
-    return;
-  }
-  if (INF_SIZE_UOP_CACHE || INF_SIZE_UOP_CACHE_PW_SIZE_LIM) {
-    init_hash_table(&inf_size_uop_cache, "infinite sized uop cache", 15000000, sizeof(int));
-  }
-
-  // The cache library computes the number of entries from cache_size_bytes/cache_line_size_bytes,
-  cpp_cache_create(UOP_CACHE_NAME, UOP_CACHE_LINES, UOP_CACHE_ASSOC, UOP_CACHE_LINE_SIZE,
-                   UOP_CACHE_REPL, /*tag_incl_offset=*/TRUE, UOP_CACHE_LINE_DATA_SIZE);
-}
-
-Flag pw_insert(Uop_Cache_Data pw) {
-  Uop_Cache_Data* cur_line_data = NULL;
-  Addr line_addr;
-  static Addr* evicted_pws = NULL;
-  UNUSED(line_addr);
-  pw.used = 0;
-
-  if (evicted_pws == NULL)
-    evicted_pws = (Addr*)calloc(UOP_CACHE_ASSOC, sizeof(Addr));
-
-  int lines_needed = pw.n_uops / ISSUE_WIDTH;
-  if (pw.n_uops % ISSUE_WIDTH) lines_needed++;
-  ASSERT(0, lines_needed > 0);
-
-  if (UOP_CACHE_INSERT_ONLY_ONPATH && pw.first_op_offpath) {
-    STAT_EVENT(ic->proc_id, UOP_CACHE_PW_INSERT_FAILED_OFFPATH);
-    return FALSE;
-  } else if (lines_needed > UOP_CACHE_ASSOC) {  // Is the PW too big?
-    STAT_EVENT(ic->proc_id, UOP_CACHE_PW_INSERT_FAILED_TOO_LONG);
-    return FALSE;
-  } else if (cpp_cache_access(UOP_CACHE_NAME, pw.first, FALSE, FALSE)) {
-    STAT_EVENT(ic->proc_id, UOP_CACHE_PW_INSERT_FAILED_CACHE_HIT);
-    return FALSE;
-  } else {
-    memset(evicted_pws, 0, UOP_CACHE_ASSOC*sizeof(*evicted_pws));
-    cur_line_data = (Uop_Cache_Data*) cpp_cache_insert(UOP_CACHE_NAME, pw.first, lines_needed,
-                                                       0, evicted_pws);
-    *cur_line_data = pw;
-    for (int i = 0; i < UOP_CACHE_ASSOC && evicted_pws[i]; i++) {
-      DEBUG(ic->proc_id, "Evicted PW: %llx\n", evicted_pws[i]);
-    }
-    DEBUG(ic->proc_id,
-          "PW inserted. off_path=%u, addr=0x%llx, set=%u, lines_needed=%i, ",
-          pw.first_op_offpath, pw.first,
-          cpp_cache_index(UOP_CACHE_NAME, pw.first, &line_addr, &line_addr),
-          lines_needed);
-    STAT_EVENT(0, UOP_CACHE_PWS_INSERTED_ONPATH + pw.first_op_offpath);
-    INC_STAT_EVENT(0, UOP_CACHE_LINES_INSERTED_ONPATH + pw.first_op_offpath, lines_needed);
-  }
-  return TRUE;
-}
-
-/**************************************************************************************/
-/* insert_uop_cache: private method, only called by accumulate_op
- *                   Drain buffer and insert. Return whether inserted.
- */
-Flag insert_uop_cache() {
-  // PW may span multiple cache entries. 1 entry per line. Additional terminating conditions per line:
-  // 1. max uops per line
-  // 2. max imm/disp per line
-  // 3. max micro-coded instr per line (not simulated)
-
-  // Invalidate after hitting maximum allowed number of lines, as in gem5
-  // Each entry/pw indexed by physical addr of 1st instr, so insert each line using same addr
-  // Invalidation: Let LRU handle it by placing PW in one line at a time.   
-  
-  ASSERT(0, accumulating_pw.n_uops);
-  Flag success = FALSE;
-  Flag new_entry;
-
-  int lines_needed = accumulating_pw.n_uops / ISSUE_WIDTH;
-  if (accumulating_pw.n_uops % ISSUE_WIDTH) lines_needed++;
-
-  if (INF_SIZE_UOP_CACHE || (INF_SIZE_UOP_CACHE_PW_SIZE_LIM 
-      && accumulating_pw.n_uops <= INF_SIZE_UOP_CACHE_PW_SIZE_LIM)) {
-    for (int ii = 0; ii < accumulating_pw.n_uops; ii++) {
-      Op* op = uop_q[ii];
-      if (!op->off_path)
-        hash_table_access_create(&inf_size_uop_cache, op->inst_info->addr, 
-                                 &new_entry);
-    }
-    success = TRUE;
-  } else if (INF_SIZE_UOP_CACHE_PW_SIZE_LIM 
-            && accumulating_pw.n_uops > INF_SIZE_UOP_CACHE_PW_SIZE_LIM) {
-    success = FALSE;
-  } else if (UOP_CACHE_INSERT_ONLY_AFTER_RESTEER_UOP_QUEUE_NOT_FULL
-             && !uop_cache_insert_enable) {
-    success = FALSE;
-  } else {
-    success = pw_insert(accumulating_pw);
-  }
-
-  return success;
-}
-
-Uop_Cache_Line_Data* uop_cache_lookup_line(Addr line_addr, Addr ft_start, Addr ft_length, Flag update_repl) {
-  if (!UOP_CACHE_ENABLE) {
-    return NULL;
-  }
-  return NULL;
-}
-
-static inline Flag in_uop_cache_search(Addr search_addr, Flag update_repl, Flag offpath) {
-  static Uop_Cache_Data cur_pw = {0};
-  Addr line_addr;
-  UNUSED(line_addr);
-  Uop_Cache_Data* uoc_data = NULL;
-
-  // First check if current pw has this search addr
-  if (cur_pw.first && search_addr >= cur_pw.first
-                   && search_addr <= cur_pw.last) {
-    uoc_data = &cur_pw;
-    if (uoc_data)
-      DEBUG(ic->proc_id, "UOC hit (cur_pw). addr=0x%llx, set=%u, update_repl=%u\n",
-            search_addr, cpp_cache_index(UOP_CACHE_NAME, search_addr, &line_addr, &line_addr),
-            update_repl);
-  } else {
-    // Next try to access a new PW starting at this addr
-    uoc_data = cpp_cache_access(UOP_CACHE_NAME, search_addr, update_repl, 0);
-    // Only update state if this access should change state
-    if (update_repl && uoc_data) {
-      if (!uoc_data->used && !offpath) {
-        STAT_EVENT(0, UOP_CACHE_LINES_INSERTED_ONPATH_USED_ONPATH + uoc_data->first_op_offpath);
-      }
-      uoc_data->used += 1;
-      cur_pw = *uoc_data;
-      DEBUG(ic->proc_id, "UOC hit (new PW). addr=0x%llx, set=%u\n",
-            search_addr, cpp_cache_index(UOP_CACHE_NAME, search_addr, &line_addr, &line_addr));
-    } else if (update_repl) {
-      memset(&cur_pw, 0, sizeof(cur_pw));
-      DEBUG(ic->proc_id, "UOC miss. addr=0x%llx, set=%u\n",  
-            search_addr, cpp_cache_index(UOP_CACHE_NAME, search_addr, &line_addr, &line_addr));
-      // maybe also add uop granularity, to make sure ending conditions are OK
-    }
-  }
-
-  return uoc_data != NULL;
-}
-
-/**************************************************************************************/
-/* in_uop_cache: Iterate over all possible ops and check if contained.
- *                    Other option: use cache to simulate capacity and maintain a map 
- *                      data structure for pcs
- */
-Flag in_uop_cache(Addr pc, Flag update_repl, Flag offpath) {
-  // A PW can span multiple cache entries. The next line used is either physically 
-  // next in the set (use flag) or anywhere in the set (use pointer), depending on impl.
-  // Here, don't care about order, just search all lines for pc in question
-  if (!UOP_CACHE_ON) {
-    return FALSE;
-  }
-
-  if (ORACLE_PERFECT_UOP_CACHE) {
-    if (update_repl) {
-      STAT_EVENT(0, UOP_CACHE_HIT);
-      STAT_EVENT(ic->proc_id, UOP_CACHE_HIT_ONPATH + offpath);
-    }
-    return TRUE;
-  } else if (INF_SIZE_UOP_CACHE || INF_SIZE_UOP_CACHE_PW_SIZE_LIM) {
-    return hash_table_access(&inf_size_uop_cache, pc) != NULL;
-  }
-
-  Flag found = in_uop_cache_search(pc, update_repl, offpath);
-  if (update_repl) {
-    STAT_EVENT(0, UOP_CACHE_MISS + found);
-    STAT_EVENT(ic->proc_id, UOP_CACHE_MISS_ONPATH + 2*found + offpath);
-  }
-  
-  return found;
-}
-
-void end_accumulate(void) {
-  if (!UOP_CACHE_ON) {
-    return;
-  }
-
-  if (accumulating_pw.n_uops > 0) {
-    Stat_Enum stat_to_inc = UOP_CACHE_PW_LENGTH_1 + (accumulating_pw.n_uops-1);
-    if (stat_to_inc > UOP_CACHE_PW_LENGTH_MAX) stat_to_inc = UOP_CACHE_PW_LENGTH_MAX;
-    STAT_EVENT(ic->proc_id, stat_to_inc);
-    insert_uop_cache();
-    memset(&accumulating_pw, 0, sizeof(accumulating_pw));
-  }
-}
-
-/**************************************************************************************/
-/* accumulate_op: accumulate into buffer. Insert into cache at end of PW. */
-void accumulate_op(Op* op) {
-  // Prediction Window termination conditions:
-  // 1. end of icache line
-  // 2. branch predicted taken
-  // 3. predetermined number of branch NT (no limit in implementation)
-  // 4. uop queue full
-  // 5. too many uops to fit in entire set, even after evicting all entries
-
-  // It is possible for an instr to be partially in 2 lines;
-  // for pw termination purposes, assume it is in first line.
-  if (!UOP_CACHE_ON) {
-    return;
-  }
-
-  Addr cur_icache_line_addr = get_cache_line_addr(&ic->icache,
-                                                  accumulating_pw.first);
-  Addr icache_line_addr = get_cache_line_addr(&ic->icache, op->inst_info->addr);
-
-  // Skipped ops means that there was a uop cache hit, so the accumulating_pw
-  // may need to be flushed. This must be done in the same place accumulation is done.
-  if (op->op_num > next_accum_op) {
-    STAT_EVENT(0, UOP_CACHE_ICACHE_SWITCH_ONPATH + op->off_path);
-    end_accumulate();
-    cur_icache_line_addr = 0;
-  }
-  ASSERT(proc_id, op->op_num >= next_accum_op);
-  if (!cur_icache_line_addr) {
-    accumulating_pw.first = op->inst_info->addr;
-    accumulating_pw.first_op_offpath = op->off_path;
-    cur_icache_line_addr = icache_line_addr;
-    next_accum_op = op->op_num + 1;
-  } else {
-    ASSERT(0, op->op_num == next_accum_op);
-    next_accum_op++;
-  }
-  DEBUG(proc_id, "next_accum_op updated to %llu\n", next_accum_op);
-  
-  Flag end_of_icache_line = icache_line_addr != cur_icache_line_addr;
-  Flag uop_q_full = (accumulating_pw.n_uops + 1 == UOP_QUEUE_SIZE);
-  if (end_of_icache_line || uop_q_full) {
-    end_accumulate();
-  }
-
-  if (accumulating_pw.n_uops == 0) {
-    // occurs when THIS fxn call drains the uop queue (insertion into uop cache)
-    accumulating_pw.first = op->inst_info->addr;
-  }
-  uop_q[accumulating_pw.n_uops] = op;
-  accumulating_pw.last = op->inst_info->addr;
-  accumulating_pw.last_op_num = op->op_num;
-  accumulating_pw.n_uops++;
-
-  // PW should be terminated at predicted-taken, but we terminate if actually-taken.
-  Flag branch_taken = op->table_info->cf_type && op->oracle_info.dir == TAKEN;
-  if (branch_taken) {
-    end_accumulate();
-  }
-};
-
-void recover_uop_cache(void) {
-  if (accumulating_pw.last_op_num > bp_recovery_info->recovery_op_num) {
-    memset(&accumulating_pw, 0, sizeof(accumulating_pw));
-  } else {
-    end_accumulate();
-  }
-  // Ops may be decoded out of order: e.g. when a previous op is slowly getting 
-  // decoded, and the current op is (speculatively) fetched already-decoded from the uop cache.
-  // So, ops preceding the recovering op may not have called accumulate_op yet.
-  if (next_accum_op > bp_recovery_info->recovery_op_num) {
-    next_accum_op = bp_recovery_info->recovery_op_num + 1;
-    DEBUG(proc_id, "UOC recovery. next_accum_op reset to %llu\n", next_accum_op);
-  }
-  if (UOP_CACHE_INSERT_ONLY_AFTER_RESTEER_UOP_QUEUE_NOT_FULL && get_uop_queue_stage_length() < UOP_QUEUE_LENGTH) {
-    set_uop_cache_insert_enable(TRUE);
-  }
-}
-
-void set_uop_cache_insert_enable(Flag new_val) {
-  uop_cache_insert_enable = new_val;
-}
diff --git a/src/uop_cache.cc b/src/uop_cache.cc
new file mode 100644
index 0000000..ea1d894
--- /dev/null
+++ b/src/uop_cache.cc
@@ -0,0 +1,482 @@
+/***************************************************************************************
+ * File         : uop_cache.h
+ * Author       : Peter Braun
+ * Date         : 10.28.2020
+ * Description  : Interface for interacting with uop cache object.
+ *                  Following Kotra et. al.'s MICRO 2020 description of uop cache baseline
+ ***************************************************************************************/
+
+#include "debug/debug_macros.h"
+#include "debug/debug_print.h"
+#include "globals/assert.h"
+#include "globals/global_defs.h"
+#include "globals/global_types.h"
+#include "globals/global_vars.h"
+#include "globals/utils.h"
+#include "isa/isa_macros.h"
+
+#include "bp/bp.h"
+#include "op_pool.h"
+
+#include "core.param.h"
+#include "debug/debug.param.h"
+#include "general.param.h"
+#include "statistics.h"
+
+#include "libs/cache_lib.h"
+#include "memory/memory.h"
+#include "memory/memory.param.h"
+#include "libs/cpp_cache.h"
+#include "uop_cache.h"
+#include "icache_stage.h"
+#include "uop_queue_stage.h"
+
+#include <vector>
+/**************************************************************************************/
+/* Macros */
+
+#define DEBUG(proc_id, args...) _DEBUG(proc_id, DEBUG_UOP_CACHE, ##args)
+
+// Uop cache is byte-addressable, so tag/set index are generated from full address (no offset)
+// Uop cache uses icache tag + icache offset as full TAG
+#define UOP_CACHE_LINE_SIZE       ICACHE_LINE_SIZE
+
+typedef std::pair<Addr, FT_Info_Static> Uop_Cache_Key;
+/**************************************************************************************/
+/* Local Prototypes */
+
+// uop cache derived class
+class Uop_Cache: public Cpp_Cache<Uop_Cache_Key, Uop_Cache_Data> {
+ protected:
+  uns   offset_bits;
+  uns64 set_mask;
+
+  // implementing the virtual hash function
+  uns set_idx_hash(Uop_Cache_Key key);
+
+ public:
+  // constructor
+  Uop_Cache(uns nl, uns asc, uns lb, Repl_Policy rp) : Cpp_Cache<Uop_Cache_Key, Uop_Cache_Data>::Cpp_Cache(nl, asc, lb, rp) {
+    offset_bits  = LOG2(line_bytes);
+    set_mask     = N_BIT_MASK(LOG2(num_sets)) << offset_bits;
+  }
+};
+
+uns Uop_Cache::set_idx_hash(Uop_Cache_Key key) {
+  return (key.first & set_mask) >> offset_bits;
+}
+
+// overload operator == of FT_Info_Static type
+bool operator==(const FT_Info_Static& lhs, const FT_Info_Static& rhs) {
+  return lhs.start == rhs.start && lhs.length == rhs.length && lhs.n_uops == rhs.n_uops;
+}
+
+// overload operator == of Uop_Cache_Key type
+bool operator==(const Uop_Cache_Key& lhs, const Uop_Cache_Key& rhs) {
+  bool pc_match = lhs.first == rhs.first;
+
+  FT_Info_Static l_info = lhs.second;
+  FT_Info_Static r_info = rhs.second;
+
+  return pc_match && l_info == r_info;
+}
+
+/**************************************************************************************/
+/* Global Variables */
+
+uns8 uop_cache_proc_id;
+// per core caches
+std::vector<Uop_Cache*> per_core_uop_cache;
+
+// uop cache per core accumulation structures
+// the accumulation buffer stores the uop cache lines of an FT
+// accumulated at the end of decoding pipeline.
+// all lines are inserted into the uop cache when the entire FT has been accumulated
+static std::vector<std::vector<Uop_Cache_Data>> per_core_accumulation_buffer;
+static std::vector<uns> per_core_num_accumulated_lines;
+static std::vector<Uop_Cache_Data> per_core_accumulating_line;
+static std::vector<FT_Info> per_core_accumulating_ft;
+static std::vector<Counter> per_core_accumulating_op_num;
+// pointers to the structures of the current core in use
+static std::vector<Uop_Cache_Data>* current_accumulation_buffer = NULL;
+static uns* current_num_accumulated_lines = NULL;
+static Uop_Cache_Data* current_accumulating_line = NULL;
+static FT_Info* current_accumulating_ft = NULL;
+static Counter* current_accumulating_op_num = NULL;
+
+// uop cache per core lookup structures
+// the lookup buffer stores the uop cache lines of an FT
+// to be consumed by the icache stage.
+// all lines are cleared when the entire FT has been consumed by the icache stage
+static std::vector<std::vector<Uop_Cache_Data>> per_core_lookup_buffer;
+static std::vector<uns> per_core_num_looked_up_lines;
+// pointers to the structures of the current core in use
+static std::vector<Uop_Cache_Data>* current_lookup_buffer = NULL;
+static uns* current_num_looked_up_lines = NULL;
+
+void alloc_mem_uop_cache(uns num_cores) {
+  if (!UOP_CACHE_ENABLE) {
+    return;
+  }
+  per_core_accumulation_buffer.resize(num_cores);
+  for (uns i = 0; i < num_cores; i++) {
+    per_core_accumulation_buffer[i].resize(UOP_CACHE_ASSOC);
+  }
+  per_core_num_accumulated_lines.resize(num_cores);
+  per_core_accumulating_line.resize(num_cores);
+  per_core_accumulating_ft.resize(num_cores);
+  per_core_accumulating_op_num.resize(num_cores);
+
+  per_core_lookup_buffer.resize(num_cores);
+  for (uns i = 0; i < num_cores; i++) {
+    per_core_lookup_buffer[i].resize(UOP_CACHE_ASSOC);
+  }
+  per_core_num_looked_up_lines.resize(num_cores);
+
+  per_core_uop_cache.resize(num_cores);
+}
+/**************************************************************************************/
+/* init_uop_cache */
+
+void init_uop_cache(uns8 proc_id) {
+  if (!UOP_CACHE_ENABLE) {
+    return;
+  }
+
+  // The cache library computes the number of entries from cache_size_bytes/cache_line_size_bytes,
+  per_core_uop_cache[proc_id] = new Uop_Cache(UOP_CACHE_LINES, UOP_CACHE_ASSOC, UOP_CACHE_LINE_SIZE,
+                                              (Repl_Policy)UOP_CACHE_REPL);
+}
+
+void set_uop_cache(uns8 proc_id) {
+  if (!UOP_CACHE_ENABLE) {
+    return;
+  }
+
+  uop_cache_proc_id = proc_id;
+
+  current_accumulation_buffer = &per_core_accumulation_buffer[proc_id];
+  current_num_accumulated_lines = &per_core_num_accumulated_lines[proc_id];
+  current_accumulating_line = &per_core_accumulating_line[proc_id];
+  current_accumulating_ft = &per_core_accumulating_ft[proc_id];
+  current_accumulating_op_num = &per_core_accumulating_op_num[proc_id];
+
+  current_lookup_buffer = &per_core_lookup_buffer[proc_id];
+  current_num_looked_up_lines = &per_core_num_looked_up_lines[proc_id];
+}
+
+Flag uop_cache_lookup_ft_and_fill_lookup_buffer(FT_Info ft_info, Flag offpath) {
+  if (!UOP_CACHE_ENABLE) {
+    return FALSE;
+  }
+
+  Uop_Cache_Data* uoc_data = NULL;
+  Addr lookup_addr = ft_info.static_info.start;
+  int buffer_index = 0;
+  do {
+    uoc_data = uop_cache_lookup_line(lookup_addr, ft_info, TRUE);
+    if (buffer_index == 0) {
+      DEBUG(uop_cache_proc_id, "UOC %s. ft_start=0x%llx, ft_length=%lld\n",
+            uoc_data ? "hit" : "miss", ft_info.static_info.start, ft_info.static_info.length);
+      if (!uoc_data) {
+        return FALSE;
+      } else {
+        if (!uoc_data->used && !offpath) {
+          STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERTED_ONPATH_USED_ONPATH + uoc_data->ft_info_dynamic.first_op_off_path);
+        }
+      }
+    }
+
+    ASSERT(uop_cache_proc_id, uoc_data);
+
+    if (!uoc_data->used && !offpath) {
+      STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERTED_ONPATH_USED_ONPATH + uoc_data->ft_info_dynamic.first_op_off_path);
+      uoc_data->used += 1;
+    }
+
+    current_lookup_buffer->at(buffer_index) = *uoc_data;
+    buffer_index++;
+    ASSERT(uop_cache_proc_id, (uoc_data->offset == 0) == uoc_data->end_of_ft);
+    lookup_addr += uoc_data->offset;
+  } while (!uoc_data->end_of_ft);
+
+  return TRUE;
+}
+
+Uop_Cache_Data* uop_cache_get_line_from_lookup_buffer() {
+  Uop_Cache_Data* uop_cache_line = &current_lookup_buffer->at(*current_num_looked_up_lines);
+  *current_num_looked_up_lines += 1;
+  return uop_cache_line;
+}
+
+void uop_cache_clear_lookup_buffer() {
+  if (!UOP_CACHE_ENABLE) {
+    return;
+  }
+
+  std::fill(current_lookup_buffer->begin(), current_lookup_buffer->end(), Uop_Cache_Data{});
+  *current_num_looked_up_lines = 0;
+}
+
+Uop_Cache_Data* uop_cache_lookup_line(Addr line_start, FT_Info ft_info, Flag update_repl) {
+  if (!UOP_CACHE_ENABLE) {
+    return NULL;
+  }
+
+  Uop_Cache_Data* uoc_data = per_core_uop_cache[uop_cache_proc_id]->access({line_start, ft_info.static_info}, update_repl == TRUE);
+  return uoc_data;
+}
+
+void clear_accumulation(Flag clear_all) {
+  *current_accumulating_line = {};
+
+  if (clear_all) {
+    std::fill(current_accumulation_buffer->begin(), current_accumulation_buffer->end(), Uop_Cache_Data{});
+    *current_num_accumulated_lines = 0;
+    *current_accumulating_ft = {};
+    *current_accumulating_op_num = 0;
+  }
+}
+
+/**************************************************************************************/
+/* end_line_accumulate: called when a uop cache line is built. */
+/* the line is added to the buffer, and when the FT has ended, */
+/* the entire buffer is inserted into the uop cache. */
+/* inserting an FT entirely avoids corner cases, but is not the most accurate. */
+void end_line_accumulate(Flag last_line_of_ft) {
+  if (!UOP_CACHE_ENABLE) {
+    return;
+  }
+
+  // add line to the buffer
+  if (*current_num_accumulated_lines < UOP_CACHE_ASSOC) {
+    current_accumulation_buffer->at(*current_num_accumulated_lines) = *current_accumulating_line;
+  }
+  *current_num_accumulated_lines += 1;
+
+  if (last_line_of_ft) {
+    // corner case handling
+    // if an inst generates more uops than the uop cache line width,
+    // consecutive lines might share the same start address (indicated by a zero offset).
+    // it causes ambiguity and we do not insert the FT.
+    Flag ft_contains_zero_offset_line = FALSE;
+    for (uns i = 0; i < *current_num_accumulated_lines; i++) {
+      Uop_Cache_Data* insert_line = &current_accumulation_buffer->at(i);
+      if (!insert_line->end_of_ft && insert_line->offset == 0) {
+        ft_contains_zero_offset_line = TRUE;
+        break;
+      }
+    }
+
+    // insert accumulation buffer
+    if (UOP_CACHE_INSERT_ONLY_ONPATH && current_accumulating_ft->dynamic_info.first_op_off_path) {
+      // if asked to only insert on-path FTs and the current FT is off-path,
+      // no insertion for the entire FT
+    } else if (ft_contains_zero_offset_line) {
+      // the corner case described above; do not insert
+      Uop_Cache_Data* uop_cache_line = uop_cache_lookup_line(current_accumulating_ft->static_info.start, *current_accumulating_ft, FALSE);
+      ASSERT(uop_cache_proc_id, !uop_cache_line);
+      if (current_accumulating_ft->dynamic_info.first_op_off_path) {
+        STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERT_FAILED_INST_TOO_BIG_OFF_PATH);
+        INC_STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERT_FAILED_INST_TOO_BIG_OFF_PATH, *current_num_accumulated_lines);
+      } else {
+        STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERT_FAILED_INST_TOO_BIG_ON_PATH);
+        INC_STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERT_FAILED_INST_TOO_BIG_ON_PATH, *current_num_accumulated_lines);
+      }
+    } else if (*current_num_accumulated_lines > UOP_CACHE_ASSOC) {
+      // if the FT is too big, do not insert
+      Uop_Cache_Data* uop_cache_line = uop_cache_lookup_line(current_accumulating_ft->static_info.start, *current_accumulating_ft, FALSE);
+      ASSERT(uop_cache_proc_id, !uop_cache_line);
+      if (current_accumulating_ft->dynamic_info.first_op_off_path) {
+        STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERT_FAILED_FT_TOO_BIG_OFF_PATH);
+        INC_STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERT_FAILED_FT_TOO_BIG_OFF_PATH, *current_num_accumulated_lines);
+      } else {
+        STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERT_FAILED_FT_TOO_BIG_ON_PATH);
+        INC_STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERT_FAILED_FT_TOO_BIG_ON_PATH, *current_num_accumulated_lines);
+      }
+    } else {
+      bool lines_exist = false;
+      for (uns i = 0; i < *current_num_accumulated_lines; i++) {
+        Uop_Cache_Data* insert_line = &current_accumulation_buffer->at(i);
+        Uop_Cache_Data* uop_cache_line = uop_cache_lookup_line(insert_line->line_start, *current_accumulating_ft, TRUE);
+
+        if (i == 0 && uop_cache_line) {
+          // when i == 0, we are looking up the first line from the FT;
+          // if this is already in the uop cache, all lines of the FT should be in the uop cache;
+          // otherwise, all lines should not be in the uop cache.
+          lines_exist = true;
+        }
+
+        if (lines_exist) {
+          // this line could already exist in the uop cache if
+          // the reuse distance in cycle is too short:
+          // the first occurrence was not yet inserted while the second was looked up,
+          // so by the time the second occurrence was inserted, the first occurrence was already in the uop cache.
+          // in that case the look-up above has updated the replacement policy,
+          // and we skip the insertion.
+          ASSERT(uop_cache_proc_id, uop_cache_line);
+
+          // the line should be identical
+          ASSERT(uop_cache_proc_id, uop_cache_line->n_uops == insert_line->n_uops &&
+                                    uop_cache_line->offset == insert_line->offset &&
+                                    uop_cache_line->end_of_ft == insert_line->end_of_ft);
+
+          if (current_accumulating_ft->dynamic_info.first_op_off_path) {
+            if (i == 0) {
+              STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERT_CONFLICTED_SHORT_REUSE_OFF_PATH);
+            }
+            STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERT_CONFLICTED_SHORT_REUSE_OFF_PATH);
+          } else {
+            if (i == 0) {
+              STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERT_CONFLICTED_SHORT_REUSE_ON_PATH);
+            }
+            STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERT_CONFLICTED_SHORT_REUSE_ON_PATH);
+          }
+        } else {
+          ASSERT(uop_cache_proc_id, !uop_cache_line);
+
+          Entry<Uop_Cache_Key, Uop_Cache_Data> evicted_entry = per_core_uop_cache[uop_cache_proc_id]->insert({insert_line->line_start, current_accumulating_ft->static_info}, *insert_line);
+
+          DEBUG(uop_cache_proc_id,
+                "uop cache line inserted. off_path=%u, addr=0x%llx\n",
+                current_accumulating_ft->dynamic_info.first_op_off_path, insert_line->line_start);
+
+          if (evicted_entry.valid) {
+            // the insertion above evicted a line
+            // need to invalidate all lines from the same FT
+            FT_Info_Static evicted_ft_info_static = evicted_entry.key.second;
+            Addr invlaidate_addr = evicted_ft_info_static.start;
+            Entry<Uop_Cache_Key, Uop_Cache_Data> invalidated_entry{};
+            do {
+              invalidated_entry = per_core_uop_cache[uop_cache_proc_id]->invalidate({invlaidate_addr, evicted_ft_info_static});
+              // if the invalidation missed, it means that the line was the one evicted at first
+              if (!invalidated_entry.valid) {
+                invalidated_entry = evicted_entry;
+              }
+              invlaidate_addr += invalidated_entry.data.offset;
+            } while (!invalidated_entry.data.end_of_ft);
+          }
+
+          if (current_accumulating_ft->dynamic_info.first_op_off_path) {
+            if (i == 0) {
+              STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERT_SUCCEEDED_OFF_PATH);
+            }
+            STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERT_SUCCEEDED_OFF_PATH);
+          } else {
+            if (i == 0) {
+              STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_INSERT_SUCCEEDED_ON_PATH);
+            }
+            STAT_EVENT(uop_cache_proc_id, UOP_CACHE_LINE_INSERT_SUCCEEDED_ON_PATH);
+          }
+        }
+      }
+    }
+
+    // stats
+    if (*current_num_accumulated_lines > UOP_CACHE_FT_LINES_8_OFF_PATH - UOP_CACHE_FT_LINES_1_OFF_PATH + 1) {
+      if (current_accumulating_ft->dynamic_info.first_op_off_path) {
+        STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_LINES_9_AND_MORE_OFF_PATH);
+      } else {
+        STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_LINES_9_AND_MORE_ON_PATH);
+      }
+    } else {
+      if (current_accumulating_ft->dynamic_info.first_op_off_path) {
+        STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_LINES_1_OFF_PATH + *current_num_accumulated_lines - 1);
+      } else {
+        STAT_EVENT(uop_cache_proc_id, UOP_CACHE_FT_LINES_1_ON_PATH + *current_num_accumulated_lines - 1);
+      }
+    }
+  }
+  clear_accumulation(last_line_of_ft);
+}
+
+/**************************************************************************************/
+/* accumulate_op: accumulate into buffer. Insert into cache at end of an uop cache line. */
+void accumulate_op(Op* op) {
+  // FT may span multiple cache entries. Additional terminating conditions per line:
+  // 1. max uops per line
+  // 2. max imm/disp per line (not simulated)
+  // 3. max micro-coded instr per line (not simulated)
+
+  if (!UOP_CACHE_ENABLE) {
+    return;
+  }
+
+  // uop cache line begin detection
+  // these two conditions should be identical
+  ASSERT(uop_cache_proc_id, (*current_num_accumulated_lines == 0 && current_accumulating_line->n_uops == 0) ==
+                            (op->bom && op->inst_info->addr == op->ft_info.static_info.start));
+  if (current_accumulating_line->n_uops == 0) {
+    if (*current_num_accumulated_lines == 0) {
+      // set current FT info
+      *current_accumulating_ft = op->ft_info;
+      ASSERT(uop_cache_proc_id, current_accumulating_ft->dynamic_info.first_op_off_path == op->off_path);
+    }
+    // set per line meta info
+    current_accumulating_line->ft_info_dynamic = current_accumulating_ft->dynamic_info;
+    current_accumulating_line->line_start = op->inst_info->addr;
+  }
+
+  // once current FT is set, they should be identical within the FT
+  ASSERT(uop_cache_proc_id, current_accumulating_ft->static_info.start == op->ft_info.static_info.start &&
+                            current_accumulating_ft->static_info.length == op->ft_info.static_info.length);
+
+  current_accumulating_line->n_uops++;
+  *current_accumulating_op_num = op->op_num;
+
+  // uop cache line end detection
+  Addr addr_following_ft = current_accumulating_ft->static_info.start + current_accumulating_ft->static_info.length;
+  Addr addr_following_inst = op->inst_info->addr + op->inst_info->trace_info.inst_size;
+  // condition 1: if the FT ends
+  bool end_condition_1 = op->eom && addr_following_inst == addr_following_ft;
+  // condition 2: if the uop cache line reaches the uop num limit
+  bool end_condition_2 = current_accumulating_line->n_uops == ISSUE_WIDTH;
+  if (end_condition_1 || end_condition_2) {
+    if (end_condition_1) {
+      current_accumulating_line->end_of_ft = TRUE;
+    } else {
+      // otherwise, calculate offset pointing to the next uop cache line
+      ASSERT(uop_cache_proc_id, current_accumulating_line->line_start);
+      Addr next_line_start;
+      if (op->eom) {
+        next_line_start = op->inst_info->addr + op->inst_info->trace_info.inst_size;
+      } else {
+        next_line_start = op->inst_info->addr;
+      }
+
+      // if next_line_start != npc, the decoupled fe did not end the FT correctly (mispredict or btb miss), or op is off-path
+      ASSERT(uop_cache_proc_id, next_line_start == op->oracle_info.npc || op->oracle_info.recover_at_decode || op->oracle_info.recover_at_exec || op->off_path);
+      current_accumulating_line->offset = next_line_start - current_accumulating_line->line_start;
+    }
+
+    end_line_accumulate(end_condition_1);
+  }
+}
+
+void recover_uop_cache(void) {
+  if (!UOP_CACHE_ENABLE) {
+    return;
+  }
+
+  if (*current_num_accumulated_lines != 0 || current_accumulating_line->n_uops != 0) {
+    if (*current_accumulating_op_num >= bp_recovery_info->recovery_op_num) {
+      clear_accumulation(TRUE);
+    } else {
+      // Ops may be decoded out of order: e.g. when a previous op is slowly getting
+      // decoded, and the current op is (speculatively) fetched already-decoded from the uop cache.
+      // So, ops preceding the recovering op may not have called accumulate_op yet.
+      // should not affect the current FT accumulation.
+
+      if (bp_recovery_info->recovery_op->ft_info.static_info == current_accumulating_ft->static_info) {
+        // a FT currently accumulating is caught up in a recovery... how come?
+        // recall the corner case where the accumulated FT is already in the uop cache because the reuse distance is small.
+        // that should be the case here.
+        ASSERT(uop_cache_proc_id, uop_cache_lookup_line(current_accumulating_ft->static_info.start, *current_accumulating_ft, FALSE) != NULL);
+      }
+    }
+  } else {
+    // no accumulation on-going
+    ASSERT(uop_cache_proc_id, current_accumulating_ft->static_info == FT_Info_Static{} &&
+                              *current_accumulating_op_num == 0);
+  }
+}
\ No newline at end of file
diff --git a/src/uop_cache.h b/src/uop_cache.h
index a3e0d26..b54d06c 100644
--- a/src/uop_cache.h
+++ b/src/uop_cache.h
@@ -11,41 +11,47 @@
 #ifndef __UOP_CACHE_H__
 #define __UOP_CACHE_H__
 
-#include "op.h"
+#ifdef __cplusplus
+extern "C" {
+#endif
 
-/**************************************************************************************/
-/* Macros */
-#define UOP_CACHE_LATENCY 1  // Simulating a uop cache latency > 1 is unsupported
+#include "op.h"
 
-typedef struct Uop_Cache_Line_Data_Struct {
+// Uop Cache Data
+typedef struct Uop_Cache_Data_struct {
+  Addr line_start;
   // number of uops in this line
-  int num_uops;
+  uns n_uops;
   // the offset for calculating the next line
   Addr offset;
+  FT_Info_Dynamic ft_info_dynamic;
   // is this line the end of the FT?
-  Flag end;
-} Uop_Cache_Line_Data;
+  Flag end_of_ft;
+
+  Counter used;
+  Flag priority;
+} Uop_Cache_Data;
 
 /**************************************************************************************/
 /* Prototypes */
 
-// only one instance of uop cache
-
+void alloc_mem_uop_cache(uns num_cores);
 void init_uop_cache(uns8 proc_id);
+void set_uop_cache(uns8 proc_id);
 void recover_uop_cache(void);
 
-Uop_Cache_Line_Data* uop_cache_lookup_line(Addr line_addr, Addr ft_start, Addr ft_length, Flag update_repl);
-
-/* return whether the instr pc is cached (this does not consider that the whole PW 
-    could already have been fetched, potentially introducing 1 incorrect cycle of latency)*/
-Flag in_uop_cache(Addr pc, Flag update_repl, Flag offpath); 
+Flag uop_cache_lookup_ft_and_fill_lookup_buffer(FT_Info ft_info, Flag offpath);
+Uop_Cache_Data* uop_cache_get_line_from_lookup_buffer(void);
+void uop_cache_clear_lookup_buffer(void);
+Uop_Cache_Data* uop_cache_lookup_line(Addr line_start, FT_Info ft_info, Flag update_repl);
 
-void end_accumulate(void);
+void clear_accumulation(Flag clear_line_only);
+void end_line_accumulate(Flag last_line_of_ft);
 /* accumulate uop into buffer. If terminating condition reached, call insert_uop_cache */
 void accumulate_op(Op* op);
 
-Flag pw_insert(Uop_Cache_Data pw);
-void set_uop_cache_insert_enable(Flag new_val);
-Uop_Cache_Data get_pw_lookahead_buffer(Addr addr);
+#ifdef __cplusplus
+}
+#endif
 
 #endif /* #ifndef __UOP_CACHE_H__ */
diff --git a/src/uop_queue_stage.cc b/src/uop_queue_stage.cc
index ecc3002..5658a5b 100644
--- a/src/uop_queue_stage.cc
+++ b/src/uop_queue_stage.cc
@@ -173,8 +173,4 @@ void update_uop_queue_fill_time_stat() {
       *new_unique_pw_entry = unique_pws_since_recovery;
     }
   }
-  if (UOP_CACHE_INSERT_ONLY_AFTER_RESTEER_UOP_QUEUE_NOT_FULL
-      && q.size() == UOP_QUEUE_LENGTH) {
-    set_uop_cache_insert_enable(FALSE);
-  }
 }
